// github/lib/views/branch-menu-view.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _desc, _value, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _coreDecorators = require('core-decorators');

var _commands = require('./commands');

var _commands2 = _interopRequireDefault(_commands);

var _gitShellOutStrategy = require('../git-shell-out-strategy');

var _propTypes3 = require('../prop-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

let BranchMenuView = (_class = (_temp = _class2 = class BranchMenuView extends _react2.default.Component {

  constructor(props, context) {
    super(props, context);

    this.state = {
      createNew: false,
      checkingOutBranch: null
    };
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.repository !== nextProps.repository) {
      this.setState({ checkingOutBranch: null });
    }
  }

  render() {
    const currentBranchName = this.state.checkingOutBranch || this.props.currentBranch.getName();
    const branchNames = this.props.branches.map(branch => branch.getName());
    if (this.state.checkingOutBranch && branchNames.indexOf(this.state.checkingOutBranch) === -1) {
      branchNames.push(this.state.checkingOutBranch);
    }

    const newBranchEditor = _react2.default.createElement(
      'div',
      { className: 'github-BranchMenuView-item github-BranchMenuView-editor' },
      _react2.default.createElement('atom-text-editor', {
        ref: e => {
          this.editorElement = e;
        },
        mini: true,
        softWrapped: true,
        placeholderText: 'enter new branch name',
        lineNumberGutterVisible: false,
        showInvisibles: false,
        scrollPastEnd: false
      })
    );

    const selectBranchView = _react2.default.createElement(
      'select',
      {
        className: 'github-BranchMenuView-item github-BranchMenuView-select input-select',
        onChange: this.didSelectItem,
        value: currentBranchName },
      this.props.currentBranch.isDetached() && _react2.default.createElement(
        'option',
        { key: 'detached', value: 'detached', disabled: true },
        this.props.currentBranch.getName()
      ),
      branchNames.map(branchName => {
        return _react2.default.createElement(
          'option',
          { key: branchName, value: branchName },
          branchName
        );
      })
    );

    return _react2.default.createElement(
      'div',
      { className: 'github-BranchMenuView' },
      _react2.default.createElement(
        _commands2.default,
        { registry: this.props.commandRegistry, target: '.github-BranchMenuView-editor atom-text-editor[mini]' },
        _react2.default.createElement(_commands.Command, { command: 'tool-panel:unfocus', callback: this.cancelCreateNewBranch }),
        _react2.default.createElement(_commands.Command, { command: 'core:cancel', callback: this.cancelCreateNewBranch }),
        _react2.default.createElement(_commands.Command, { command: 'core:confirm', callback: this.createBranch })
      ),
      _react2.default.createElement(
        'div',
        { className: 'github-BranchMenuView-selector' },
        _react2.default.createElement('span', { className: 'github-BranchMenuView-item icon icon-git-branch' }),
        this.state.createNew ? newBranchEditor : selectBranchView,
        _react2.default.createElement(
          'button',
          { className: 'github-BranchMenuView-item github-BranchMenuView-button btn',
            onClick: this.createBranch },
          ' New Branch '
        )
      )
    );
  }

  didSelectItem(event) {
    var _this = this;

    return _asyncToGenerator(function* () {
      const branchName = event.target.value;
      _this.setState({ checkingOutBranch: branchName });

      try {
        yield _this.props.checkout(branchName);
      } catch (e) {
        _this.setState({ checkingOutBranch: null });
        if (!(e instanceof _gitShellOutStrategy.GitError)) {
          throw e;
        }
        if (e.stdErr.match(/local changes.*would be overwritten/)) {
          const files = e.stdErr.split(/\r?\n/).filter(function (l) {
            return l.startsWith('\t');
          }).map(function (l) {
            return `\`${l.trim()}\``;
          }).join('<br>');
          _this.props.notificationManager.addError('Checkout aborted', {
            description: 'Local changes to the following would be overwritten:<br>' + files + '<br>Please commit your changes or stash them.',
            dismissable: true
          });
        } else {
          _this.props.notificationManager.addError('Checkout aborted', { description: e.stdErr, dismissable: true });
        }
      }
      return null;
    })();
  }

  createBranch() {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      if (_this2.state.createNew) {
        const branchName = _this2.editorElement.getModel().getText().trim();
        try {
          yield _this2.props.checkout(branchName, { createNew: true });
          _this2.setState({ createNew: false, checkingOutBranch: branchName });
        } catch (e) {
          if (!(e instanceof _gitShellOutStrategy.GitError)) {
            throw e;
          }
          if (e.stdErr.match(/branch.*already exists/)) {
            _this2.props.notificationManager.addError('Cannot create branch', {
              description: `\`${branchName}\` already exists. Choose another branch name.`
            });
          } else if (e.stdErr.match(/error: you need to resolve your current index first/)) {
            _this2.props.notificationManager.addError('Cannot create branch', {
              description: 'You must first resolve merge conflicts.',
              dismissable: true
            });
          } else {
            _this2.props.notificationManager.addError('Cannot create branch', { description: e.stdErr, dismissable: true });
          }
          _this2.setState({ createNew: false });
        }
      } else {
        _this2.setState({ createNew: true }, function () {
          return _this2.editorElement.focus();
        });
      }
    })();
  }

  cancelCreateNewBranch() {
    this.setState({ createNew: false });
  }
}, _class2.propTypes = {
  workspace: _propTypes2.default.object.isRequired,
  commandRegistry: _propTypes2.default.object.isRequired,
  notificationManager: _propTypes2.default.object.isRequired,
  repository: _propTypes2.default.object,
  branches: _propTypes2.default.arrayOf(_propTypes3.BranchPropType).isRequired,
  currentBranch: _propTypes3.BranchPropType.isRequired,
  checkout: _propTypes2.default.func
}, _class2.defaultProps = {
  checkout: () => Promise.resolve()
}, _temp), (_applyDecoratedDescriptor(_class.prototype, 'didSelectItem', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'didSelectItem'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'createBranch', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'createBranch'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'cancelCreateNewBranch', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'cancelCreateNewBranch'), _class.prototype)), _class);
exports.default = BranchMenuView;
