// github/lib/views/changed-files-count-view.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _octicon = require('./octicon');

var _octicon2 = _interopRequireDefault(_octicon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let ChangedFilesCountView = (_temp = _class = class ChangedFilesCountView extends _react2.default.Component {

  render() {
    const label = this.props.changedFilesCount === 1 ? '1 file' : `${this.props.changedFilesCount} files`;
    return _react2.default.createElement(
      'a',
      {
        ref: 'changedFiles',
        className: 'github-ChangedFilesCount inline-block icon icon-diff',
        onClick: this.props.didClick },
      label,
      this.props.mergeConflictsPresent && _react2.default.createElement(_octicon2.default, { icon: 'alert' })
    );
  }
}, _class.propTypes = {
  changedFilesCount: _propTypes2.default.number.isRequired,
  didClick: _propTypes2.default.func.isRequired,
  mergeConflictsPresent: _propTypes2.default.bool
}, _class.defaultProps = {
  changedFilesCount: 0,
  mergeConflictsPresent: false,
  didClick: () => {}
}, _temp);
exports.default = ChangedFilesCountView;
