// github/lib/views/commands.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Command = exports.default = undefined;

var _class, _temp, _class2, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _propTypes3 = require('../prop-types');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let Commands = (_temp = _class = class Commands extends _react2.default.Component {

  render() {
    var _props = this.props;
    const registry = _props.registry,
          target = _props.target;

    return _react2.default.createElement(
      'div',
      null,
      _react2.default.Children.map(this.props.children, child => {
        return _react2.default.cloneElement(child, { registry, target });
      })
    );
  }
}, _class.propTypes = {
  registry: _propTypes2.default.object.isRequired,
  target: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes3.DOMNodePropType]).isRequired,
  children: _propTypes2.default.oneOfType([_propTypes2.default.element, _propTypes2.default.arrayOf(_propTypes2.default.element)]).isRequired
}, _temp);
exports.default = Commands;
let Command = exports.Command = (_temp2 = _class2 = class Command extends _react2.default.Component {

  componentDidMount() {
    this.registerCommand(this.props);
  }

  componentWillReceiveProps(newProps) {
    for (const prop of ['registry', 'target', 'command', 'callback']) {
      if (newProps[prop] !== this.props[prop]) {
        this.disposable.dispose();
        this.registerCommand(newProps);
      }
    }
  }

  componentWillUnmount() {
    this.disposable.dispose();
  }

  registerCommand(_ref) {
    let registry = _ref.registry,
        target = _ref.target,
        command = _ref.command,
        callback = _ref.callback;

    this.disposable = registry.add(target, command, callback);
  }

  render() {
    return null;
  }
}, _class2.propTypes = {
  registry: _propTypes2.default.object,
  target: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes3.DOMNodePropType]),
  command: _propTypes2.default.string.isRequired,
  callback: _propTypes2.default.func.isRequired
}, _temp2);
