// github/lib/views/debugger-view.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _etch = require('etch');

var _etch2 = _interopRequireDefault(_etch);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } /** @jsx etch.dom */
/* eslint react/no-unknown-property: "off" */

const defaultStyle = {
  position: 'fixed',
  zIndex: 100000000,
  backgroundColor: 'white',
  minWidth: '300px',
  minHeight: '300px',
  top: '400px',
  left: '600px',
  maxWidth: '800px',
  maxHeight: '400px',
  overflow: 'auto',
  whiteSpace: 'pre',
  fontFamily: 'monospace',
  border: '3px solid black'
};

let DebuggerView = class DebuggerView {
  constructor(props) {
    this.props = props;
    _etch2.default.initialize(this);
  }

  update(props) {
    this.props = props;
    _etch2.default.update(this);
  }

  render() {
    var _props = this.props;

    const data = _props.data,
          style = _props.style,
          others = _objectWithoutProperties(_props, ['data', 'style']);

    const finalStyle = _extends({}, defaultStyle, style);

    return _etch2.default.dom(
      'div',
      _extends({ style: finalStyle }, others),
      JSON.stringify(data, null, '  ')
    );
  }
};
exports.default = DebuggerView;
