// github/lib/views/github-login-view.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _desc, _value, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _coreDecorators = require('core-decorators');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

let GithubLoginView = (_class = (_temp = _class2 = class GithubLoginView extends _react2.default.Component {

  constructor(props, context) {
    super(props, context);
    this.state = {
      loggingIn: false,
      token: ''
    };
  }

  render() {
    let subview;
    if (this.state.loggingIn) {
      subview = this.renderTokenInput();
    } else {
      subview = this.renderLogin();
    }

    return _react2.default.createElement(
      'div',
      { className: 'github-GithubLoginView' },
      subview
    );
  }

  renderLogin() {
    return _react2.default.createElement(
      'div',
      { className: 'github-GithubLoginView-Subview' },
      this.props.children,
      _react2.default.createElement(
        'button',
        { onClick: this.handleLoginClick, className: 'btn btn-primary btn-lg icon icon-octoface' },
        'Login'
      )
    );
  }

  renderTokenInput() {
    return _react2.default.createElement(
      'form',
      { className: 'github-GithubLoginView-Subview', onSubmit: this.handleSubmitToken },
      _react2.default.createElement(
        'p',
        null,
        'Step 1: Visit ',
        _react2.default.createElement(
          'a',
          { href: 'https://github.atom.io/login' },
          'github.atom.io/login'
        ),
        ' to generate an authentication token.'
      ),
      _react2.default.createElement(
        'p',
        null,
        'Step 2: Enter the token below:'
      ),
      _react2.default.createElement('input', {
        type: 'text',
        className: 'input-text native-key-bindings',
        placeholder: 'Enter your token...',
        value: this.state.token,
        onChange: this.handleTokenChange
      }),
      _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          'button',
          { type: 'button', onClick: this.handleCancelTokenClick, className: 'btn icon inline-block-tight' },
          'Cancel'
        ),
        _react2.default.createElement('input', {
          type: 'submit',
          value: 'Login',
          onClick: this.handleSubmitTokenClick, className: 'btn btn-primary icon icon-check inline-block-tight'
        })
      )
    );
  }

  handleLoginClick() {
    this.setState({ loggingIn: true });
  }

  handleCancelTokenClick(e) {
    e.preventDefault();
    this.setState({ loggingIn: false });
  }

  handleSubmitTokenClick(e) {
    e.preventDefault();
    this.handleSubmitToken();
  }

  handleSubmitToken() {
    this.props.onLogin(this.state.token);
  }

  handleTokenChange(e) {
    this.setState({ token: e.target.value });
  }
}, _class2.propTypes = {
  children: _propTypes2.default.node,
  onLogin: _propTypes2.default.func
}, _class2.defaultProps = {
  children: _react2.default.createElement(
    'p',
    null,
    'Log in to GitHub to access PR information and more!'
  ),
  onLogin: token => {}
}, _temp), (_applyDecoratedDescriptor(_class.prototype, 'handleLoginClick', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleLoginClick'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'handleCancelTokenClick', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleCancelTokenClick'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'handleSubmitTokenClick', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleSubmitTokenClick'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'handleSubmitToken', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleSubmitToken'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'handleTokenChange', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleTokenChange'), _class.prototype)), _class);
exports.default = GithubLoginView;
