// github/lib/views/list-view.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _etch = require('etch');

var _etch2 = _interopRequireDefault(_etch);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } /** @jsx etch.dom */
/* eslint react/no-unknown-property: "off" */

let ListView = class ListView {
  constructor(props) {
    this.props = props;
    _etch2.default.initialize(this);
  }

  didClickItem(e, item) {
    if (e.detail === 1) {
      if (this.props.didSelectItem) {
        this.props.didSelectItem(item);
        return _etch2.default.update(this);
      }
    } else if (e.detail === 2) {
      if (this.props.didConfirmItem) {
        this.props.didConfirmItem(item);
      }
    }
    return null;
  }

  update(props) {
    this.props = props;
    return _etch2.default.update(this);
  }

  render() {
    // eslint-disable-next-line no-unused-vars
    var _props = this.props;

    const ref = _props.ref,
          didSelectItem = _props.didSelectItem,
          didConfirmItem = _props.didConfirmItem,
          items = _props.items,
          selectedItems = _props.selectedItems,
          renderItem = _props.renderItem,
          others = _objectWithoutProperties(_props, ['ref', 'didSelectItem', 'didConfirmItem', 'items', 'selectedItems', 'renderItem']);

    return _etch2.default.dom(
      'div',
      others,
      items.map((item, index) => renderItem(item, selectedItems.has(item), e => this.didClickItem(e, item)))
    );
  }

  destroy() {
    _etch2.default.destroy(this);
  }
};
exports.default = ListView;
