// github/lib/views/portal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * `Portal` is a mechanism for rendering a React subtree at a different place
 * in the DOM.
 *
 *    <Portal type="span" className="portal-class" appendNode={true}>
 *      <Stuff />
 *    </Portal>
 *
 * Given the above example, there will be a span with the class "portal-class"
 * created and appended to the document body, and then `<Stuff />` will be
 * rendered into it. Note that this uses `unstable_renderSubtreeIntoContainer`
 * to preserve context in the subtree.
 *
 * `getElement()` allows access to the React subtree container element.
 * `getRenderedSubtree()` allows access to the rendered subtree instance
 * (`Stuff` in the example above).
 *
 * Pass `false` (the default) to `appendNode` to skip adding the node to the
 * DOM. `type` defaults to "div" and `className` defaults to
 * "react-atom-portal".
 */
let Portal = (_temp = _class = class Portal extends _react2.default.Component {

  componentDidMount() {
    let node;
    if (this.props.getDOMNode) {
      node = this.props.getDOMNode();
    }

    if (!node) {
      node = document.createElement(this.props.type);
      node.className = this.props.className;
    }

    this.node = node;

    if (this.props.appendNode) {
      document.body.appendChild(this.node);
    }
    this.renderPortal(this.props);
  }

  componentWillReceiveProps(newProps) {
    this.renderPortal(newProps);
  }

  componentWillUnmount() {
    _reactDom2.default.unmountComponentAtNode(this.node);
    if (this.props.appendNode) {
      document.body.removeChild(this.node);
    }
  }

  renderPortal(props) {
    this.subtree = _reactDom2.default.unstable_renderSubtreeIntoContainer(this, props.children, this.node);
  }

  shouldComponentUpdate() {
    return false;
  }

  render() {
    return null;
  }

  getRenderedSubtree() {
    return this.subtree;
  }

  getElement() {
    return this.node;
  }

  getView() {
    if (this.view) {
      return this.view;
    }

    const override = {
      getPortal: () => this,
      getInstance: () => this.subtree,
      getElement: this.getElement.bind(this)
    };

    this.view = new Proxy(override, {
      get(target, name) {
        if (Reflect.has(target, name)) {
          return target[name];
        }

        return target.getInstance()[name];
      },

      set(target, name, value) {
        target.getInstance()[name] = value;
      },

      has(target, name) {
        return Reflect.has(target.getInstance(), name) || Reflect.has(target, name);
      }
    });
    return this.view;
  }
}, _class.propTypes = {
  type: _propTypes2.default.string,
  className: _propTypes2.default.string,
  appendNode: _propTypes2.default.bool,
  getDOMNode: _propTypes2.default.func
}, _class.defaultProps = {
  type: 'div',
  className: 'react-atom-portal',
  appendNode: false,
  getDOMNode: null
}, _temp);
exports.default = Portal;
