// github/lib/views/pr-url-input-box.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _desc, _value, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _coreDecorators = require('core-decorators');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

let PrUrlInputBox = (_class = (_temp = _class2 = class PrUrlInputBox extends _react2.default.Component {

  constructor(props, context) {
    super(props, context);
    this.state = {
      url: ''
    };
  }

  render() {
    return _react2.default.createElement(
      'form',
      { className: 'github-PrUrlInputBox-Subview', onSubmit: this.handleSubmitUrl },
      this.props.children,
      _react2.default.createElement('input', {
        type: 'text',
        className: 'input-text native-key-bindings',
        placeholder: 'e.g. https://github.com/owner/repo/pull/123',
        value: this.state.url,
        onChange: this.handleUrlChange
      }),
      _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement('input', {
          type: 'submit',
          value: 'Submit',
          onClick: this.handleSubmitUrlClick, className: 'btn btn-primary icon icon-check inline-block-tight'
        })
      )
    );
  }

  handleSubmitUrlClick(e) {
    e.preventDefault();
    this.handleSubmitUrl();
  }

  handleSubmitUrl() {
    this.props.onSubmit(this.state.url);
  }

  handleUrlChange(e) {
    this.setState({ url: e.target.value });
  }
}, _class2.propTypes = {
  onSubmit: _propTypes2.default.func.isRequired,
  children: _propTypes2.default.node
}, _temp), (_applyDecoratedDescriptor(_class.prototype, 'handleSubmitUrlClick', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleSubmitUrlClick'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'handleSubmitUrl', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleSubmitUrl'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'handleUrlChange', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleUrlChange'), _class.prototype)), _class);
exports.default = PrUrlInputBox;
