// github/lib/views/push-pull-view.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let PushPullView = (_temp = _class = class PushPullView extends _react2.default.Component {

  render() {
    const pushing = this.props.pushInProgress;
    const pulling = this.props.fetchInProgress;
    const pushClasses = (0, _classnames2.default)('github-PushPull-icon', 'icon', { 'icon-arrow-up': !pushing, 'icon-sync': pushing });
    const pullClasses = (0, _classnames2.default)('github-PushPull-icon', 'icon', { 'icon-arrow-down': !pulling, 'icon-sync': pulling });
    return _react2.default.createElement(
      'div',
      { className: 'github-PushPull inline-block', ref: e => {
          this.element = e;
        } },
      _react2.default.createElement('span', { className: pullClasses }),
      _react2.default.createElement(
        'span',
        { className: 'github-PushPull-label is-pull' },
        this.props.behindCount ? `${this.props.behindCount}` : ''
      ),
      _react2.default.createElement('span', { className: pushClasses }),
      _react2.default.createElement(
        'span',
        { className: 'github-PushPull-label is-push' },
        this.props.aheadCount ? `${this.props.aheadCount}` : ''
      )
    );
  }
}, _class.propTypes = {
  pushInProgress: _propTypes2.default.bool,
  fetchInProgress: _propTypes2.default.bool,
  behindCount: _propTypes2.default.number,
  aheadCount: _propTypes2.default.number
}, _class.defaultProps = {
  pushInProgress: false,
  fetchInProgress: false,
  behindCount: 0,
  aheadCount: 0
}, _temp);
exports.default = PushPullView;
