// github/lib/views/resizer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _desc, _value, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _coreDecorators = require('core-decorators');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

let Resizer = (_class = (_temp = _class2 = class Resizer extends _react2.default.Component {

  constructor(props, context) {
    super(props, context);
    this.state = {
      dragging: false,
      start: null,
      lastSize: props.size
    };
  }

  render() {
    var _props = this.props;

    const onChange = _props.onChange,
          children = _props.children,
          className = _props.className,
          others = _objectWithoutProperties(_props, ['onChange', 'children', 'className']); // eslint-disable-line no-unused-vars


    return _react2.default.createElement(
      'div',
      _extends({}, others, { className: (0, _classnames2.default)('resizer-container', className) }),
      _react2.default.createElement(
        'div',
        { className: 'sizing-handle', onMouseDown: this.handleMouseDown },
        '\xA0'
      ),
      _react2.default.createElement(
        'div',
        { className: 'content', style: { width: this.props.size } },
        children
      )
    );
  }

  handleMouseDown(e) {
    // save off copy since synthetic event will be recycled
    // by the time the `setState` function runs
    const clientX = e.clientX;
    this.setState(s => ({
      dragging: true,
      start: clientX,
      lastSize: this.props.size
    }));

    document.addEventListener('mousemove', this.handleMouseMove);
    document.addEventListener('mouseup', this.handleMouseUp);
  }

  handleMouseUp(e) {
    document.removeEventListener('mousemove', this.handleMouseMove);
    document.removeEventListener('mouseup', this.handleMouseUp);

    this.setState({
      dragging: false
    });
  }

  handleMouseMove(e) {
    if (!this.state.dragging) {
      return;
    }
    const oldPos = this.state.start;
    const delta = oldPos - e.clientX;
    const newSize = Math.max(0, this.state.lastSize + delta);
    this.props.onChange(newSize);
  }
}, _class2.propTypes = {
  size: _propTypes2.default.number.isRequired,
  onChange: _propTypes2.default.func.isRequired,
  className: _propTypes2.default.string,
  children: _propTypes2.default.element
}, _temp), (_applyDecoratedDescriptor(_class.prototype, 'handleMouseDown', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleMouseDown'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'handleMouseUp', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleMouseUp'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'handleMouseMove', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleMouseMove'), _class.prototype)), _class);
exports.default = Resizer;
