// github/lib/views/simple-tooltip.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let SimpleTooltip = (_temp = _class = class SimpleTooltip extends _react2.default.Component {

  componentDidMount() {
    this.disposable = this.props.tooltips.add(_reactDom2.default.findDOMNode(this.child), { title: () => this.props.title });
  }

  componentWillUnmount() {
    this.disposable.dispose();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.title !== this.props.title) {
      this.disposable.dispose();
      this.disposable = this.props.tooltips.add(_reactDom2.default.findDOMNode(this.child), { title: () => this.props.title });
    }
  }

  render() {
    const child = _react2.default.Children.only(this.props.children);
    return _react2.default.cloneElement(child, { ref: e => {
        this.child = e;
      } });
  }
}, _class.propTypes = {
  tooltips: _propTypes2.default.object.isRequired,
  children: _propTypes2.default.node.isRequired,
  title: _propTypes2.default.string.isRequired
}, _temp);
exports.default = SimpleTooltip;
