// github/lib/views/status-bar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// TODO: updates, unmount
let StatusBar = (_temp = _class = class StatusBar extends _react2.default.Component {

  componentDidMount() {
    this.consumeStatusBar();
  }

  componentDidUpdate() {
    this.consumeStatusBar();
  }

  render() {
    return _react2.default.createElement(
      'div',
      { ref: c => {
          this.container = c;
        } },
      this.props.children
    );
  }

  consumeStatusBar() {
    if (this.tile) {
      return;
    }
    if (!this.props.statusBar) {
      return;
    }

    const componentElement = this.container.children[0];
    this.tile = this.props.statusBar.addRightTile({ item: componentElement, priority: -50 });
    this.props.onConsumeStatusBar(this.props.statusBar);
  }
}, _class.propTypes = {
  children: _propTypes2.default.element.isRequired,
  statusBar: _propTypes2.default.object,
  onConsumeStatusBar: _propTypes2.default.func
}, _class.defaultProps = {
  onConsumeStatusBar: statusBar => {}
}, _temp);
exports.default = StatusBar;
