// github/lib/views/tabs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _desc, _value, _class, _class2, _temp, _class3, _temp2, _class4, _temp3;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _coreDecorators = require('core-decorators');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

function getClasses(staticClasses, active) {
  return (0, _classnames2.default)(...staticClasses, {
    inactive: !active,
    active
  });
}

let TabsNavigationItem = (_class = (_temp = _class2 = class TabsNavigationItem extends _react2.default.Component {

  render() {
    const classes = getClasses(['github-Tabs-NavigationItem'], this.props.active);
    return _react2.default.createElement(
      'div',
      { className: classes, onClick: this.handleClick },
      this.props.title
    );
  }

  handleClick() {
    this.props.onClick(this.props.tabIndex);
  }
}, _class2.propTypes = {
  title: _propTypes2.default.node,
  tabIndex: _propTypes2.default.number.isRequired,
  active: _propTypes2.default.bool,
  onClick: _propTypes2.default.func
}, _temp), (_applyDecoratedDescriptor(_class.prototype, 'handleClick', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleClick'), _class.prototype)), _class);
let Tabs = (_temp2 = _class3 = class Tabs extends _react2.default.Component {

  render() {
    const children = _react2.default.Children.toArray(this.props.children).filter(v => !!v);
    return _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)('github-Tabs', this.props.className), 'data-tabs-count': children.length },
      _react2.default.createElement(
        'nav',
        { className: 'github-Tabs-NavigationContainer' },
        children.map((child, i) => child && _react2.default.createElement(TabsNavigationItem, {
          key: i,
          title: child.props.title,
          active: this.props.activeIndex === i,
          tabIndex: i,
          onClick: this.props.onChange
        }))
      ),
      _react2.default.createElement(
        'section',
        { className: 'github-Tabs-PanelContainer' },
        children.map((child, i) => child && _react2.default.cloneElement(child, { active: i === this.props.activeIndex }))
      )
    );
  }
}, _class3.propTypes = {
  className: _propTypes2.default.string,
  activeIndex: _propTypes2.default.number.isRequired,
  children: _propTypes2.default.node,
  onChange: _propTypes2.default.func
}, _class3.defaultProps = {
  className: '',
  onChange: () => {}
}, _temp2);
exports.default = Tabs;


Tabs.Panel = (_temp3 = _class4 = class TabPanel extends _react2.default.Component {

  render() {
    // eslint-disable-next-line no-unused-vars
    var _props = this.props;

    const active = _props.active,
          children = _props.children,
          className = _props.className,
          title = _props.title,
          others = _objectWithoutProperties(_props, ['active', 'children', 'className', 'title']);

    return _react2.default.createElement(
      'div',
      _extends({ className: getClasses(['github-Tabs-Panel', className], this.props.active) }, others),
      this.props.children
    );
  }
}, _class4.propTypes = {
  className: _propTypes2.default.string,
  active: _propTypes2.default.bool,
  children: _propTypes2.default.node,
  title: _propTypes2.default.node
}, _class4.defaultProps = {
  className: '',
  active: false
}, _temp3);
