// grammar-selector/lib/grammar-list-view.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

/** @babel */

var _atomSelectList = require('atom-select-list');

var _atomSelectList2 = _interopRequireDefault(_atomSelectList);

var GrammarListView = (function () {
  function GrammarListView() {
    var _this = this;

    _classCallCheck(this, GrammarListView);

    this.autoDetect = { name: 'Auto Detect' };
    this.selectListView = new _atomSelectList2['default']({
      itemsClassList: ['mark-active'],
      items: [],
      filterKeyForItem: function filterKeyForItem(grammar) {
        return grammar.name;
      },
      elementForItem: function elementForItem(grammar) {
        var grammarName = grammar.name || grammar.scopeName;
        var element = document.createElement('li');
        if (grammar === _this.currentGrammar) {
          element.classList.add('active');
        }
        element.textContent = grammarName;
        element.dataset.grammar = grammarName;
        return element;
      },
      didConfirmSelection: function didConfirmSelection(grammar) {
        _this.cancel();
        if (grammar === _this.autoDetect) {
          atom.textEditors.clearGrammarOverride(_this.editor);
        } else {
          atom.textEditors.setGrammarOverride(_this.editor, grammar.scopeName);
        }
      },
      didCancelSelection: function didCancelSelection() {
        _this.cancel();
      }
    });
    this.selectListView.element.classList.add('grammar-selector');
  }

  _createClass(GrammarListView, [{
    key: 'destroy',
    value: function destroy() {
      this.cancel();
      return this.selectListView.destroy();
    }
  }, {
    key: 'cancel',
    value: function cancel() {
      if (this.panel != null) {
        this.panel.destroy();
      }
      this.panel = null;
      this.currentGrammar = null;
      if (this.previouslyFocusedElement) {
        this.previouslyFocusedElement.focus();
        this.previouslyFocusedElement = null;
      }
    }
  }, {
    key: 'attach',
    value: function attach() {
      this.previouslyFocusedElement = document.activeElement;
      if (this.panel == null) {
        this.panel = atom.workspace.addModalPanel({ item: this.selectListView });
      }
      this.selectListView.focus();
      this.selectListView.reset();
    }
  }, {
    key: 'toggle',
    value: _asyncToGenerator(function* () {
      if (this.panel != null) {
        this.cancel();
      } else if (atom.workspace.getActiveTextEditor()) {
        this.editor = atom.workspace.getActiveTextEditor();
        this.currentGrammar = this.editor.getGrammar();
        if (this.currentGrammar === atom.grammars.nullGrammar) {
          this.currentGrammar = this.autoDetect;
        }

        var grammars = atom.grammars.getGrammars().filter(function (grammar) {
          return grammar !== atom.grammars.nullGrammar;
        });
        grammars.sort(function (a, b) {
          if (a.scopeName === 'text.plain') {
            return -1;
          } else if (b.scopeName === 'text.plain') {
            return 1;
          } else if (a.name) {
            return a.name.localeCompare(b.name);
          } else if (a.scopeName) {
            return a.scopeName.localeCompare(b.scopeName);
          } else {
            return 1;
          }
        });
        grammars.unshift(this.autoDetect);
        yield this.selectListView.update({ items: grammars });
        this.attach();
      }
    })
  }]);

  return GrammarListView;
})();

exports['default'] = GrammarListView;
module.exports = exports['default'];
