// autosave/lib/autosave.js
(function() {
  var CompositeDisposable, Disposable, dontSaveIf, fs, ref, ref1, shouldSave;

  ref = require('atom'), CompositeDisposable = ref.CompositeDisposable, Disposable = ref.Disposable;

  fs = require('fs-plus');

  ref1 = require('./controls'), dontSaveIf = ref1.dontSaveIf, shouldSave = ref1.shouldSave;

  module.exports = {
    subscriptions: null,
    provideService: function() {
      return {
        dontSaveIf: dontSaveIf
      };
    },
    activate: function() {
      var handleBeforeUnload, handleBlur;
      this.subscriptions = new CompositeDisposable;
      handleBeforeUnload = this.autosaveAllPaneItems.bind(this);
      window.addEventListener('beforeunload', handleBeforeUnload, true);
      this.subscriptions.add(new Disposable(function() {
        return window.removeEventListener('beforeunload', handleBeforeUnload, true);
      }));
      handleBlur = (function(_this) {
        return function(event) {
          if (event.target === window) {
            return _this.autosaveAllPaneItems();
          } else if (event.target.matches('atom-text-editor:not(mini)') && !event.target.contains(event.relatedTarget)) {
            return _this.autosavePaneItem(event.target.getModel());
          }
        };
      })(this);
      window.addEventListener('blur', handleBlur, true);
      this.subscriptions.add(new Disposable(function() {
        return window.removeEventListener('blur', handleBlur, true);
      }));
      return this.subscriptions.add(atom.workspace.onWillDestroyPaneItem((function(_this) {
        return function(arg) {
          var item;
          item = arg.item;
          return _this.autosavePaneItem(item);
        };
      })(this)));
    },
    deactivate: function() {
      return this.subscriptions.dispose();
    },
    autosavePaneItem: function(paneItem) {
      var pane;
      if (!atom.config.get('autosave.enabled')) {
        return;
      }
      if ((paneItem != null ? typeof paneItem.getURI === "function" ? paneItem.getURI() : void 0 : void 0) == null) {
        return;
      }
      if (!(paneItem != null ? typeof paneItem.isModified === "function" ? paneItem.isModified() : void 0 : void 0)) {
        return;
      }
      if (!(((paneItem != null ? typeof paneItem.getPath === "function" ? paneItem.getPath() : void 0 : void 0) != null) && fs.isFileSync(paneItem.getPath()))) {
        return;
      }
      if (!shouldSave(paneItem)) {
        return;
      }
      pane = atom.workspace.paneForItem(paneItem);
      if (pane != null) {
        return pane.saveItem(paneItem);
      } else {
        return typeof paneItem.save === "function" ? paneItem.save() : void 0;
      }
    },
    autosaveAllPaneItems: function() {
      var i, len, paneItem, ref2, results;
      ref2 = atom.workspace.getPaneItems();
      results = [];
      for (i = 0, len = ref2.length; i < len; i++) {
        paneItem = ref2[i];
        results.push(this.autosavePaneItem(paneItem));
      }
      return results;
    }
  };

}).call(this);
