// github/lib/controllers/git-tab-controller.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _desc, _value, _class, _class2, _temp; /** @jsx etch.dom */
/* eslint react/no-unknown-property: "off" */

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _etch = require('etch');

var _etch2 = _interopRequireDefault(_etch);

var _eventKit = require('event-kit');

var _gitTabView = require('../views/git-tab-view');

var _gitTabView2 = _interopRequireDefault(_gitTabView);

var _modelObserver = require('../models/model-observer');

var _modelObserver2 = _interopRequireDefault(_modelObserver);

var _branch = require('../models/branch');

var _commit = require('../models/commit');

var _coreDecorators = require('core-decorators');

var _yubikiri = require('yubikiri');

var _yubikiri2 = _interopRequireDefault(_yubikiri);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

let GitTabController = (_class = (_temp = _class2 = class GitTabController {

  constructor(props) {
    this.props = props;
    this.stagingOperationInProgress = false;
    this.lastFocus = _gitTabView2.default.focus.STAGING;

    this.repositoryObserver = new _modelObserver2.default({
      fetchData: this.fetchRepositoryData,
      didUpdate: () => {
        this.refreshResolutionProgress(false, false);
        return this.update();
      }
    });
    this.repositoryObserver.setActiveModel(props.repository);
    _etch2.default.initialize(this);

    this.element.addEventListener('focusin', this.rememberLastFocus);
    this.subscriptions = new _eventKit.Disposable(() => this.element.removeEventListener('focusin', this.rememberLastFocus));
  }

  serialize() {
    return {
      deserializer: 'GitTabControllerStub'
    };
  }

  render() {
    const modelData = this.repositoryObserver.getActiveModelData() || this.defaultRepositoryData();
    const hasUndoHistory = this.props.repository ? this.hasUndoHistory() : false;
    return _etch2.default.dom(_gitTabView2.default, _extends({
      ref: 'gitTab'
    }, modelData, {
      repository: this.props.repository,
      resolutionProgress: this.props.resolutionProgress,
      workspace: this.props.workspace,
      commandRegistry: this.props.commandRegistry,
      notificationManager: this.props.notificationManager,
      initializeRepo: this.props.initializeRepo,
      didSelectFilePath: this.props.didSelectFilePath,
      didDiveIntoFilePath: this.props.didDiveIntoFilePath,
      didSelectMergeConflictFile: this.props.didSelectMergeConflictFile,
      didDiveIntoMergeConflictPath: this.props.didDiveIntoMergeConflictPath,
      focusFilePatchView: this.props.focusFilePatchView,
      stageFilePatch: this.stageFilePatch,
      unstageFilePatch: this.unstageFilePatch,
      openFiles: this.props.openFiles,
      discardWorkDirChangesForPaths: this.props.discardWorkDirChangesForPaths,
      attemptFileStageOperation: this.attemptFileStageOperation,
      prepareToCommit: this.prepareToCommit,
      commit: this.commit,
      setAmending: this.setAmending,
      isAmending: this.props.isAmending,
      hasUndoHistory: hasUndoHistory,
      undoLastDiscard: this.props.undoLastDiscard,
      abortMerge: this.abortMerge,
      resolveAsOurs: this.resolveAsOurs,
      resolveAsTheirs: this.resolveAsTheirs,
      push: this.push,
      pull: this.pull,
      fetch: this.fetch,
      checkout: this.checkout
    }));
  }

  update(props) {
    var _this = this;

    return _asyncToGenerator(function* () {
      const oldProps = _this.props;
      _this.props = _extends({}, _this.props, props);
      if (_this.props.repository !== oldProps.repository) {
        yield _this.repositoryObserver.setActiveModel(props.repository);
      } else if (_this.props.isAmending !== oldProps.isAmending) {
        yield _this.repositoryObserver.refreshModelData(_this.getActiveRepository());
      }
      return _etch2.default.update(_this);
    })();
  }

  destroy() {
    this.subscriptions.dispose();
    this.repositoryObserver.destroy();
  }

  getTitle() {
    return 'Git';
  }

  getIconName() {
    return 'git-commit';
  }

  getDefaultLocation() {
    return 'right';
  }

  getPreferredWidth() {
    return 400;
  }

  getURI() {
    return 'atom-github://stub-uri/git-tab-controller';
  }

  getWorkingDirectory() {
    return this.props.repository.getWorkingDirectoryPath();
  }

  getLastModelDataRefreshPromise() {
    return this.repositoryObserver.getLastModelDataRefreshPromise();
  }

  getActiveRepository() {
    return this.repositoryObserver.getActiveModel();
  }

  refreshModelData() {
    return this.repositoryObserver.refreshModelData();
  }

  fetchRepositoryData(repository) {
    return (0, _yubikiri2.default)({
      lastCommit: repository.getLastCommit(),
      isMerging: repository.isMerging(),
      isRebasing: repository.isRebasing(),
      currentBranch: repository.getCurrentBranch(),
      unstagedChanges: repository.getUnstagedChanges(),
      stagedChanges: this.fetchStagedChanges(repository),
      mergeConflicts: repository.getMergeConflicts(),
      workingDirectoryPath: repository.getWorkingDirectoryPath(),
      mergeMessage: (() => {
        var _ref = _asyncToGenerator(function* (query) {
          const isMerging = yield query.isMerging;
          return isMerging ? repository.getMergeMessage() : null;
        });

        return function mergeMessage(_x) {
          return _ref.apply(this, arguments);
        };
      })(),
      fetchInProgress: Promise.resolve(false)
    });
  }

  defaultRepositoryData() {
    return {
      lastCommit: _commit.nullCommit,
      isMerging: false,
      isRebasing: false,
      currentBranch: _branch.nullBranch,
      unstagedChanges: [],
      stagedChanges: [],
      mergeConflicts: [],
      workingDirectoryPath: this.props.repository.getWorkingDirectoryPath(),
      mergeMessage: null,
      fetchInProgress: true
    };
  }

  fetchStagedChanges(repository) {
    if (this.props.isAmending) {
      return repository.getStagedChangesSinceParentCommit();
    } else {
      return repository.getStagedChanges();
    }
  }

  /*
   * Begin (but don't await) an async conflict-counting task for each merge conflict path that has no conflict
   * marker count yet. Omit any path that's already open in a TextEditor or that has already been counted.
   *
   * includeOpen - update marker counts for files that are currently open in TextEditors
   * includeCounts - update marker counts for files that have been counted before
   */
  refreshResolutionProgress(includeOpen, includeCounted) {
    const data = this.repositoryObserver.getActiveModelData();
    if (!data) {
      return;
    }

    const openPaths = new Set(this.props.workspace.getTextEditors().map(editor => editor.getPath()));
    for (let i = 0; i < data.mergeConflicts.length; i++) {
      const conflictPath = _path2.default.join(data.workingDirectoryPath, data.mergeConflicts[i].filePath);

      if (!includeOpen && openPaths.has(conflictPath)) {
        continue;
      }

      if (!includeCounted && this.props.resolutionProgress.getRemaining(conflictPath) !== undefined) {
        continue;
      }

      this.props.refreshResolutionProgress(conflictPath);
    }
  }

  unstageFilePatch(filePatch) {
    return this.getActiveRepository().applyPatchToIndex(filePatch.getUnstagePatch());
  }

  attemptFileStageOperation(filePaths, stageStatus) {
    if (this.stagingOperationInProgress) {
      return {
        stageOperationPromise: Promise.resolve(),
        selectionUpdatePromise: Promise.resolve()
      };
    }

    this.stagingOperationInProgress = true;

    const fileListUpdatePromise = this.refs.gitTab.refs.stagingView.getNextListUpdatePromise();
    let stageOperationPromise;
    if (stageStatus === 'staged') {
      stageOperationPromise = this.unstageFiles(filePaths);
    } else {
      stageOperationPromise = this.stageFiles(filePaths);
    }
    const selectionUpdatePromise = fileListUpdatePromise.then(() => {
      this.stagingOperationInProgress = false;
    });

    return { stageOperationPromise, selectionUpdatePromise };
  }

  stageFiles(filePaths) {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      const pathsToIgnore = [];
      const repository = _this2.getActiveRepository();
      for (const filePath of filePaths) {
        if (yield repository.pathHasMergeMarkers(filePath)) {
          // eslint-disable-line no-await-in-loop
          const choice = atom.confirm({
            message: 'File contains merge markers: ',
            detailedMessage: `Do you still want to stage this file?\n${filePath}`,
            buttons: ['Stage', 'Cancel']
          });
          if (choice !== 0) {
            pathsToIgnore.push(filePath);
          }
        }
      }
      const pathsToStage = filePaths.filter(function (filePath) {
        return !pathsToIgnore.includes(filePath);
      });
      return repository.stageFiles(pathsToStage);
    })();
  }

  unstageFiles(filePaths) {
    const repository = this.getActiveRepository();
    if (this.props.isAmending) {
      return repository.stageFilesFromParentCommit(filePaths);
    } else {
      return repository.unstageFiles(filePaths);
    }
  }

  prepareToCommit() {
    var _this3 = this;

    return _asyncToGenerator(function* () {
      return !(yield _this3.props.ensureGitTab());
    })();
  }

  commit(message) {
    var _this4 = this;

    return _asyncToGenerator(function* () {
      try {
        yield _this4.getActiveRepository().commit(message, { amend: _this4.props.isAmending });
        _this4.setAmending(false);
      } catch (e) {
        if (e.code === 'ECONFLICT') {
          _this4.props.notificationManager.addError('Cannot commit without resolving all the merge conflicts first.', { dismissable: true });
        } else {
          throw e;
        }
      }
    })();
  }

  setAmending(isAmending) {
    this.props.didChangeAmending(isAmending);
  }

  abortMerge() {
    var _this5 = this;

    return _asyncToGenerator(function* () {
      const choice = atom.confirm({
        message: 'Abort merge',
        detailedMessage: 'Are you sure?',
        buttons: ['Abort', 'Cancel']
      });
      if (choice !== 0) {
        return null;
      }

      try {
        yield _this5.getActiveRepository().abortMerge();
      } catch (e) {
        if (e.code === 'EDIRTYSTAGED') {
          _this5.props.notificationManager.addError(`Cannot abort because ${e.path} is both dirty and staged.`, { dismissable: true });
        } else {
          throw e;
        }
      }
      return _etch2.default.update(_this5);
    })();
  }

  resolveAsOurs(paths) {
    var _this6 = this;

    return _asyncToGenerator(function* () {
      const data = _this6.repositoryObserver.getActiveModelData();
      if (!data) {
        return;
      }
      const side = data.isRebasing ? 'theirs' : 'ours';
      yield _this6.getActiveRepository().checkoutSide(side, paths);
      _this6.refreshResolutionProgress(false, true);
    })();
  }

  resolveAsTheirs(paths) {
    var _this7 = this;

    return _asyncToGenerator(function* () {
      const data = _this7.repositoryObserver.getActiveModelData();
      if (!data) {
        return;
      }
      const side = data.isRebasing ? 'ours' : 'theirs';
      yield _this7.getActiveRepository().checkoutSide(side, paths);
      _this7.refreshResolutionProgress(false, true);
    })();
  }

  checkout(branchName, options) {
    return this.getActiveRepository().checkout(branchName, options);
  }

  rememberLastFocus(event) {
    this.lastFocus = this.refs.gitTab.rememberFocus(event) || _gitTabView2.default.focus.STAGING;
  }

  restoreFocus() {
    this.refs.gitTab.setFocus(this.lastFocus);
  }

  hasFocus() {
    return this.element.contains(document.activeElement);
  }

  wasActivated() {
    process.nextTick(() => this.restoreFocus());
  }

  focusAndSelectStagingItem(filePath, stagingStatus) {
    return this.refs.gitTab.focusAndSelectStagingItem(filePath, stagingStatus);
  }

  quietlySelectItem(filePath, stagingStatus) {
    return this.refs.gitTab.quietlySelectItem(filePath, stagingStatus);
  }

  hasUndoHistory() {
    return this.props.repository.hasDiscardHistory();
  }
}, _class2.focus = _extends({}, _gitTabView2.default.focus), _temp), (_applyDecoratedDescriptor(_class.prototype, 'fetchRepositoryData', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'fetchRepositoryData'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'attemptFileStageOperation', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'attemptFileStageOperation'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'unstageFiles', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'unstageFiles'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'prepareToCommit', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'prepareToCommit'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'commit', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'commit'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'setAmending', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'setAmending'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'abortMerge', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'abortMerge'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'resolveAsOurs', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'resolveAsOurs'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'resolveAsTheirs', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'resolveAsTheirs'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'checkout', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'checkout'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'rememberLastFocus', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'rememberLastFocus'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'quietlySelectItem', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'quietlySelectItem'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'hasUndoHistory', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'hasUndoHistory'), _class.prototype)), _class);
exports.default = GitTabController;
