// github/lib/controllers/issueish-pane-item-controller.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _desc, _value, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _yubikiri = require('yubikiri');

var _yubikiri2 = _interopRequireDefault(_yubikiri);

var _coreDecorators = require('core-decorators');

var _relayNetworkLayerManager = require('../relay-network-layer-manager');

var _relayNetworkLayerManager2 = _interopRequireDefault(_relayNetworkLayerManager);

var _relayRootContainer = require('../containers/relay-root-container');

var _relayRootContainer2 = _interopRequireDefault(_relayRootContainer);

var _githubLoginModel = require('../models/github-login-model');

var _githubLoginModel2 = _interopRequireDefault(_githubLoginModel);

var _githubLoginView = require('../views/github-login-view');

var _githubLoginView2 = _interopRequireDefault(_githubLoginView);

var _observeModel = require('../views/observe-model');

var _observeModel2 = _interopRequireDefault(_observeModel);

var _issueishInfoByNumberRoute = require('../routes/issueish-info-by-number-route');

var _issueishInfoByNumberRoute2 = _interopRequireDefault(_issueishInfoByNumberRoute);

var _issueishLookupByNumberContainer = require('../containers/issueish-lookup-by-number-container');

var _issueishLookupByNumberContainer2 = _interopRequireDefault(_issueishLookupByNumberContainer);

var _relayEnvironment = require('../views/relay-environment');

var _relayEnvironment2 = _interopRequireDefault(_relayEnvironment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

let IssueishPaneItemController = (_class = (_temp = _class2 = class IssueishPaneItemController extends _react2.default.Component {
  fetchData(loginModel) {
    return (0, _yubikiri2.default)({
      token: loginModel.getToken(this.props.host)
    });
  }

  render() {
    return _react2.default.createElement(
      _observeModel2.default,
      { model: this.props.loginModel, fetchData: this.fetchData },
      this.renderWithToken
    );
  }

  renderWithToken(data) {
    if (!data) {
      return null;
    }
    if (data.token === _githubLoginModel.UNAUTHENTICATED) {
      return _react2.default.createElement(_githubLoginView2.default, { onLogin: this.handleLogin });
    }

    const route = new _issueishInfoByNumberRoute2.default({
      repoOwner: this.props.owner,
      repoName: this.props.repo,
      issueishNumber: this.props.issueishNumber
    });

    const environment = _relayNetworkLayerManager2.default.getEnvironmentForHost(this.props.host, data.token);

    const Component = _issueishLookupByNumberContainer2.default;
    return _react2.default.createElement(
      _relayEnvironment2.default,
      { environment: environment },
      _react2.default.createElement(_relayRootContainer2.default, {
        Component: Component,
        route: route,
        environment: environment,
        renderFetched: props => {
          return _react2.default.createElement(Component, _extends({}, props, { onTitleChange: this.props.onTitleChange }));
        },
        renderLoading: () => {
          return _react2.default.createElement(
            'div',
            { className: 'github-Loader' },
            _react2.default.createElement('span', { className: 'github-Spinner' })
          );
        },
        renderFailure: (err, retry) => {
          if (err.response && err.response.status === 401) {
            return _react2.default.createElement(
              'div',
              { className: 'github-GithubLoginView-Container' },
              _react2.default.createElement(
                _githubLoginView2.default,
                { onLogin: this.handleLogin },
                _react2.default.createElement(
                  'p',
                  null,
                  'The API endpoint returned a unauthorized error. Please try to re-authenticate with the endpoint.'
                )
              )
            );
          } else {
            return _react2.default.createElement(
              'div',
              null,
              'An unknown error occurred.'
            );
          }
        }
      })
    );
  }

  handleLogin(token) {
    this.props.loginModel.setToken(this.props.host, token);
  }
}, _class2.propTypes = {
  onTitleChange: _propTypes2.default.func.isRequired,
  owner: _propTypes2.default.string.isRequired,
  repo: _propTypes2.default.string.isRequired,
  issueishNumber: _propTypes2.default.number.isRequired,
  host: _propTypes2.default.string,
  loginModel: _propTypes2.default.instanceOf(_githubLoginModel2.default).isRequired
}, _class2.defaultProps = {
  host: 'api.github.com'
}, _temp), (_applyDecoratedDescriptor(_class.prototype, 'fetchData', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'fetchData'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'renderWithToken', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'renderWithToken'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'handleLogin', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'handleLogin'), _class.prototype)), _class);
exports.default = IssueishPaneItemController;
