// github/lib/controllers/remote-pr-controller.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _dec, _class, _desc, _value, _class2, _class3, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _coreDecorators = require('core-decorators');

var _yubikiri = require('yubikiri');

var _yubikiri2 = _interopRequireDefault(_yubikiri);

var _propTypes3 = require('../prop-types');

var _observeModel = require('../decorators/observe-model');

var _observeModel2 = _interopRequireDefault(_observeModel);

var _githubLoginView = require('../views/github-login-view');

var _githubLoginView2 = _interopRequireDefault(_githubLoginView);

var _githubLoginModel = require('../models/github-login-model');

var _remote = require('../models/remote');

var _prInfoController = require('./pr-info-controller');

var _prInfoController2 = _interopRequireDefault(_prInfoController);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

let RemotePrController = (_dec = (0, _observeModel2.default)({
  getModel: props => props.loginModel,
  fetchData: (loginModel, _ref) => {
    let host = _ref.host;

    return (0, _yubikiri2.default)({
      token: loginModel.getToken(host)
    });
  }
}), _dec(_class = (_class2 = (_temp = _class3 = class RemotePrController extends _react2.default.Component {

  render() {
    var _props = this.props;
    const host = _props.host,
          remote = _props.remote,
          currentBranchName = _props.currentBranchName,
          token = _props.token,
          loginModel = _props.loginModel,
          selectedPrUrl = _props.selectedPrUrl,
          onSelectPr = _props.onSelectPr,
          onUnpinPr = _props.onUnpinPr;

    return _react2.default.createElement(
      'div',
      { className: 'github-RemotePrController' },
      token && token !== _githubLoginModel.UNAUTHENTICATED && _react2.default.createElement(_prInfoController2.default, _extends({ host, remote, currentBranchName, token, loginModel, selectedPrUrl, onSelectPr, onUnpinPr }, {
        onLogin: this.handleLogin
      })),
      (!token || token === _githubLoginModel.UNAUTHENTICATED) && _react2.default.createElement(_githubLoginView2.default, { onLogin: this.handleLogin })
    );
  }

  handleLogin(token) {
    this.props.loginModel.setToken(this.props.host, token);
  }
}, _class3.propTypes = {
  loginModel: _propTypes2.default.object.isRequired,
  host: _propTypes2.default.string, // fully qualified URI to the API endpoint, e.g. 'https://api.github.com'
  remote: _propTypes3.RemotePropType.isRequired,
  token: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.symbol]),
  currentBranchName: _propTypes2.default.string.isRequired,
  onSelectPr: _propTypes2.default.func.isRequired,
  selectedPrUrl: _propTypes2.default.string,
  onUnpinPr: _propTypes2.default.func.isRequired
}, _class3.defaultProps = {
  host: 'https://api.github.com',
  remote: _remote.nullRemote,
  token: null
}, _temp), (_applyDecoratedDescriptor(_class2.prototype, 'handleLogin', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class2.prototype, 'handleLogin'), _class2.prototype)), _class2)) || _class);
exports.default = RemotePrController;
