// github/lib/controllers/repository-conflict-controller.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _dec, _class, _class2, _temp;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _yubikiri = require('yubikiri');

var _yubikiri2 = _interopRequireDefault(_yubikiri);

var _eventKit = require('event-kit');

var _observeModel = require('../decorators/observe-model');

var _observeModel2 = _interopRequireDefault(_observeModel);

var _resolutionProgress = require('../models/conflicts/resolution-progress');

var _resolutionProgress2 = _interopRequireDefault(_resolutionProgress);

var _editorConflictController = require('./editor-conflict-controller');

var _editorConflictController2 = _interopRequireDefault(_editorConflictController);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Render an `EditorConflictController` for each `TextEditor` open on a file that contains git conflict markers.
 */
let RepositoryConflictController = (_dec = (0, _observeModel2.default)({
  getModel: props => props.repository,
  fetchData: (r, props) => {
    return (0, _yubikiri2.default)({
      workingDirectoryPath: r.getWorkingDirectoryPath(),
      mergeConflictPaths: r.getMergeConflicts().then(conflicts => conflicts.map(conflict => conflict.filePath)),
      isRebasing: r.isRebasing()
    });
  }
}), _dec(_class = (_temp = _class2 = class RepositoryConflictController extends _react2.default.Component {

  constructor(props, context) {
    super(props, context);

    this.state = { openEditors: this.props.workspace.getTextEditors() };
    this.subscriptions = new _eventKit.CompositeDisposable();
  }

  componentDidMount() {
    this.subscriptions.add(this.props.workspace.observeTextEditors(() => {
      this.setState({
        openEditors: this.props.workspace.getTextEditors()
      });
    }));
  }

  render() {
    const conflictingEditors = this.getConflictingEditors();

    return _react2.default.createElement(
      'div',
      null,
      conflictingEditors.map(editor => _react2.default.createElement(_editorConflictController2.default, {
        key: editor.id,
        commandRegistry: this.props.commandRegistry,
        resolutionProgress: this.props.resolutionProgress,
        editor: editor,
        isRebase: this.props.isRebasing,
        refreshResolutionProgress: this.props.refreshResolutionProgress
      }))
    );
  }

  getConflictingEditors() {
    if (this.props.mergeConflictPaths.length === 0 || this.state.openEditors.length === 0) {
      return [];
    }

    const commonBasePath = this.props.repository.getWorkingDirectoryPath();
    const fullMergeConflictPaths = new Set(this.props.mergeConflictPaths.map(relativePath => _path2.default.join(commonBasePath, relativePath)));

    return this.state.openEditors.filter(editor => fullMergeConflictPaths.has(editor.getPath()));
  }

  componentWillUnmount() {
    this.subscriptions.dispose();
  }
}, _class2.propTypes = {
  workspace: _propTypes2.default.object.isRequired,
  commandRegistry: _propTypes2.default.object.isRequired,
  resolutionProgress: _propTypes2.default.object.isRequired,
  repository: _propTypes2.default.object.isRequired,
  mergeConflictPaths: _propTypes2.default.arrayOf(_propTypes2.default.string),
  isRebasing: _propTypes2.default.bool,
  refreshResolutionProgress: _propTypes2.default.func
}, _class2.defaultProps = {
  resolutionProgress: new _resolutionProgress2.default(),
  mergeConflictPaths: [],
  isRebasing: false,
  refreshResolutionProgress: () => {}
}, _temp)) || _class);
exports.default = RepositoryConflictController;
