// metrics/lib/metrics.js
(function() {
  var CompositeDisposable, IgnoredCommands, PathRE, Reporter, fs, grim, path, stripPath;

  CompositeDisposable = require('atom').CompositeDisposable;

  path = require('path');

  Reporter = require('./reporter');

  fs = require('fs-plus');

  grim = require('grim');

  IgnoredCommands = {
    'vim-mode:move-up': true,
    'vim-mode:move-down': true,
    'vim-mode:move-left': true,
    'vim-mode:move-right': true
  };

  module.exports = {
    activate: function(arg) {
      var sessionLength;
      sessionLength = arg.sessionLength;
      this.subscriptions = new CompositeDisposable;
      this.shouldIncludePanesAndCommands = Math.random() < 0.05;
      return this.ensureClientId((function(_this) {
        return function() {
          return _this.begin(sessionLength);
        };
      })(this));
    },
    deactivate: function() {
      var ref;
      return (ref = this.subscriptions) != null ? ref.dispose() : void 0;
    },
    serialize: function() {
      return {
        sessionLength: Date.now() - this.sessionStart
      };
    },
    provideReporter: function() {
      return {
        sendEvent: Reporter.sendEvent.bind(Reporter),
        sendTiming: Reporter.sendTiming.bind(Reporter),
        sendException: Reporter.sendException.bind(Reporter)
      };
    },
    begin: function(sessionLength) {
      this.sessionStart = Date.now();
      if (sessionLength) {
        Reporter.sendEvent('window', 'ended', null, sessionLength);
      }
      Reporter.sendEvent('window', 'started');
      this.subscriptions.add(atom.onDidThrowError(function(event) {
        var errorMessage;
        errorMessage = event;
        if (typeof event !== 'string') {
          errorMessage = event.message;
        }
        errorMessage = stripPath(errorMessage) || 'Unknown';
        errorMessage = errorMessage.replace('Uncaught ', '').slice(0, 150);
        return Reporter.sendException(errorMessage);
      }));
      this.subscriptions.add(atom.config.onDidChange('core.telemetryConsent', function(arg) {
        var newValue, oldValue;
        newValue = arg.newValue, oldValue = arg.oldValue;
        if (newValue !== 'undecided') {
          return Reporter.sendEvent('setting', 'core.telemetryConsent', newValue);
        }
      }));
      this.watchPaneItems();
      this.watchCommands();
      this.watchDeprecations();
      if (atom.getLoadSettings().shellLoadTime != null) {
        Reporter.sendTiming('shell', 'load', atom.getLoadSettings().shellLoadTime);
      }
      return process.nextTick(function() {
        return Reporter.sendTiming('core', 'load', atom.getWindowLoadTime());
      });
    },
    ensureClientId: function(callback) {
      if (localStorage.getItem('metrics.userId')) {
        return callback();
      } else if (atom.config.get('metrics.userId')) {
        localStorage.setItem('metrics.userId', atom.config.get('metrics.userId'));
        return callback();
      } else {
        return this.createClientId(function(clientId) {
          localStorage.setItem('metrics.userId', clientId);
          return callback();
        });
      }
    },
    createClientId: function(callback) {
      return callback(require('node-uuid').v4());
    },
    getClientId: function() {
      return localStorage.getItem('metrics.userId');
    },
    watchPaneItems: function() {
      return this.subscriptions.add(atom.workspace.onDidAddPaneItem((function(_this) {
        return function(arg) {
          var item;
          item = arg.item;
          if (!_this.shouldIncludePanesAndCommands) {
            return;
          }
          return Reporter.sendPaneItem(item);
        };
      })(this)));
    },
    watchCommands: function() {
      return this.subscriptions.add(atom.commands.onWillDispatch((function(_this) {
        return function(commandEvent) {
          var eventName, ref;
          if (!_this.shouldIncludePanesAndCommands) {
            return;
          }
          eventName = commandEvent.type;
          if ((ref = commandEvent.detail) != null ? ref.jQueryTrigger : void 0) {
            return;
          }
          if (eventName.startsWith('core:') || eventName.startsWith('editor:')) {
            return;
          }
          if (!(eventName.indexOf(':') > -1)) {
            return;
          }
          if (eventName in IgnoredCommands) {
            return;
          }
          return Reporter.sendCommand(eventName);
        };
      })(this)));
    },
    watchDeprecations: function() {
      this.deprecationCache = {};
      this.packageVersionCache = {};
      atom.packages.onDidActivateInitialPackages((function(_this) {
        return function() {
          var j, len, pack, packages, ref;
          packages = atom.packages.getLoadedPackages();
          for (j = 0, len = packages.length; j < len; j++) {
            pack = packages[j];
            _this.packageVersionCache[pack.name] = (pack != null ? (ref = pack.metadata) != null ? ref.version : void 0 : void 0) || 'unknown';
          }
          setImmediate(function() {
            var deprecation, k, len1, ref1;
            ref1 = grim.getDeprecations();
            for (k = 0, len1 = ref1.length; k < len1; k++) {
              deprecation = ref1[k];
              _this.reportDeprecation(deprecation);
            }
          });
        };
      })(this));
      atom.packages.onDidLoadPackage((function(_this) {
        return function(pack) {
          var ref;
          if (!_this.packageVersionCache[pack.name]) {
            return _this.packageVersionCache[pack.name] = (pack != null ? (ref = pack.metadata) != null ? ref.version : void 0 : void 0) || 'unknown';
          }
        };
      })(this));
      return grim.on('updated', (function(_this) {
        return function(deprecation) {
          return setImmediate(function() {
            return _this.reportDeprecation(deprecation);
          });
        };
      })(this));
    },
    reportDeprecation: function(deprecation) {
      var __, message, nameAndVersion, pack, packageName, ref, ref1, ref2, ref3, stack, version;
      message = deprecation.getMessage().slice(0, 500);
      ref = deprecation.stacks;
      for (__ in ref) {
        stack = ref[__];
        packageName = (ref1 = (ref2 = stack.metadata) != null ? ref2.packageName : void 0) != null ? ref1 : (this.getPackageName(stack) || '').toLowerCase();
        if (!packageName) {
          continue;
        }
        if (!this.packageVersionCache[packageName]) {
          pack = atom.packages.getLoadedPackage(packageName);
          this.packageVersionCache[packageName] = (pack != null ? (ref3 = pack.metadata) != null ? ref3.version : void 0 : void 0) || 'unknown';
        }
        version = this.packageVersionCache[packageName];
        nameAndVersion = packageName + "@" + version;
        if (this.deprecationCache[nameAndVersion + message] == null) {
          this.deprecationCache[nameAndVersion + message] = true;
          Reporter.sendEvent('deprecation-v3', nameAndVersion, message);
        }
      }
    },
    getFileNameFromCallSite: function(callsite) {
      var ref;
      return (ref = callsite.fileName) != null ? ref : callsite.getFileName();
    },
    getPackageName: function(stack) {
      var fileName, i, j, packageName, packagePath, packagePaths, ref, relativePath;
      packagePaths = this.getPackagePathsByPackageName();
      for (i = j = 1, ref = stack.length; 1 <= ref ? j < ref : j > ref; i = 1 <= ref ? ++j : --j) {
        fileName = this.getFileNameFromCallSite(stack[i]);
        if (!fileName) {
          return;
        }
        if (fileName.includes(path.sep + "node_modules" + path.sep)) {
          continue;
        }
        for (packageName in packagePaths) {
          packagePath = packagePaths[packageName];
          relativePath = path.relative(packagePath, fileName);
          if (!/^\.\./.test(relativePath)) {
            return packageName;
          }
        }
        if (atom.getUserInitScriptPath() === fileName) {
          return "init-script";
        }
      }
    },
    getPackagePathsByPackageName: function() {
      var j, len, pack, ref;
      if (this.packagePathsByPackageName != null) {
        return this.packagePathsByPackageName;
      }
      this.packagePathsByPackageName = {};
      ref = atom.packages.getLoadedPackages();
      for (j = 0, len = ref.length; j < len; j++) {
        pack = ref[j];
        this.packagePathsByPackageName[pack.name] = pack.path;
        if (pack.path.indexOf('.atom/dev/packages') > -1 || pack.path.indexOf('.atom/packages') > -1) {
          this.packagePathsByPackageName[pack.name] = fs.absolute(pack.path);
        }
      }
      return this.packagePathsByPackageName;
    }
  };

  PathRE = /'?((\/|\\|[a-z]:\\)[^\s']+)+'?/ig;

  stripPath = function(message) {
    return message.replace(PathRE, '<path>');
  };

}).call(this);
