// bookmarks/lib/bookmarks-view.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

/** @babel */

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _atomSelectList = require('atom-select-list');

var _atomSelectList2 = _interopRequireDefault(_atomSelectList);

var BookmarksView = (function () {
  function BookmarksView(editorsBookmarks) {
    var _this = this;

    _classCallCheck(this, BookmarksView);

    this.editorsBookmarks = editorsBookmarks;
    this.selectList = new _atomSelectList2['default']({
      emptyMessage: 'No bookmarks found',
      items: [],
      filterKeyForItem: function filterKeyForItem(bookmark) {
        return bookmark.filterText;
      },
      didConfirmSelection: function didConfirmSelection(_ref) {
        var editor = _ref.editor;
        var marker = _ref.marker;

        _this.hide();
        editor.setSelectedBufferRange(marker.getBufferRange(), { autoscroll: true });
        atom.workspace.paneForItem(editor).activate();
        atom.workspace.paneForItem(editor).activateItem(editor);
      },
      didCancelSelection: function didCancelSelection() {
        _this.hide();
      },
      elementForItem: function elementForItem(_ref2) {
        var marker = _ref2.marker;
        var editor = _ref2.editor;

        var bookmarkStartRow = marker.getStartBufferPosition().row;
        var bookmarkEndRow = marker.getEndBufferPosition().row;
        var bookmarkPath = editor.getPath() ? _path2['default'].basename(editor.getPath()) : 'untitled';
        var bookmarkLocation = bookmarkPath + ':' + (bookmarkStartRow + 1);
        if (bookmarkStartRow !== bookmarkEndRow) {
          bookmarkLocation += '-' + (bookmarkEndRow + 1);
        }

        var lineText = editor.lineTextForBufferRow(bookmarkStartRow);
        var li = document.createElement('li');
        li.classList.add('bookmark');
        var primaryLine = document.createElement('div');
        primaryLine.classList.add('primary-line');
        primaryLine.textContent = bookmarkLocation;
        li.appendChild(primaryLine);
        if (lineText) {
          var secondaryLine = document.createElement('div');
          secondaryLine.classList.add('secondary-line', 'line-text');
          secondaryLine.textContent = lineText.trim();
          li.appendChild(secondaryLine);
          li.classList.add('two-lines');
        }
        return li;
      }
    });
    this.selectList.element.classList.add('bookmarks-view');
  }

  _createClass(BookmarksView, [{
    key: 'destroy',
    value: function destroy() {
      this.selectList.destroy();
      this.getModalPanel().destroy();
      if (this.previouslyFocusedElement) {
        this.previouslyFocusedElement.focus();
        this.previouslyFocusedElement = null;
      }
    }
  }, {
    key: 'show',
    value: _asyncToGenerator(function* () {
      var bookmarks = [];
      for (var _ref32 of this.editorsBookmarks) {
        var editor = _ref32.editor;
        var markerLayer = _ref32.markerLayer;

        for (var marker of markerLayer.getMarkers()) {
          var filterText = '' + marker.getStartBufferPosition().row;

          if (editor.getPath()) {
            filterText += ' ' + editor.getPath();
          }

          var bookmarkedLineText = editor.lineTextForBufferRow(marker.getStartBufferPosition().row);
          if (bookmarkedLineText) {
            filterText += ' ' + bookmarkedLineText.trim();
          }

          bookmarks.push({ marker: marker, editor: editor, filterText: filterText });
        }
      }

      this.previouslyFocusedElement = document.activeElement;
      this.selectList.reset();
      yield this.selectList.update({ items: bookmarks });
      this.getModalPanel().show();
      this.selectList.focus();
    })
  }, {
    key: 'hide',
    value: function hide() {
      this.getModalPanel().hide();
      if (this.previouslyFocusedElement) {
        this.previouslyFocusedElement.focus();
        this.previouslyFocusedElement = null;
      }
    }
  }, {
    key: 'getModalPanel',
    value: function getModalPanel() {
      if (!this.modalPanel) {
        this.modalPanel = atom.workspace.addModalPanel({ item: this.selectList });
      }
      return this.modalPanel;
    }
  }]);

  return BookmarksView;
})();

exports['default'] = BookmarksView;
module.exports = exports['default'];
