// github/lib/models/discard-history.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

var _mkdirp = require('mkdirp');

var _mkdirp2 = _interopRequireDefault(_mkdirp);

var _discardHistoryStores = require('./discard-history-stores');

var _helpers = require('../helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const emptyFilePath = _path2.default.join(_os2.default.tmpdir(), 'empty-file.txt');
const emptyFilePromise = (0, _helpers.writeFile)(emptyFilePath, '');

let DiscardHistory = class DiscardHistory {
  constructor(createBlob, expandBlobToFile, mergeFile, workdirPath) {
    var _ref = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : {};

    let maxHistoryLength = _ref.maxHistoryLength;

    this.createBlob = createBlob;
    this.expandBlobToFile = expandBlobToFile;
    this.mergeFile = mergeFile;
    this.workdirPath = workdirPath;
    this.partialFileHistory = new _discardHistoryStores.PartialFileDiscardHistory(maxHistoryLength);
    this.wholeFileHistory = new _discardHistoryStores.WholeFileDiscardHistory(maxHistoryLength);
  }

  getLastSnapshots() {
    let partialDiscardFilePath = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

    if (partialDiscardFilePath) {
      return this.partialFileHistory.getLastSnapshotsForPath(partialDiscardFilePath);
    } else {
      return this.wholeFileHistory.getLastSnapshots();
    }
  }

  getHistory() {
    let partialDiscardFilePath = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

    if (partialDiscardFilePath) {
      return this.partialFileHistory.getHistoryForPath(partialDiscardFilePath);
    } else {
      return this.wholeFileHistory.getHistory();
    }
  }

  hasHistory() {
    let partialDiscardFilePath = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

    const history = this.getHistory(partialDiscardFilePath);
    return history.length > 0;
  }

  popHistory() {
    let partialDiscardFilePath = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

    if (partialDiscardFilePath) {
      return this.partialFileHistory.popHistoryForPath(partialDiscardFilePath);
    } else {
      return this.wholeFileHistory.popHistory();
    }
  }

  clearHistory() {
    let partialDiscardFilePath = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

    if (partialDiscardFilePath) {
      this.partialFileHistory.clearHistoryForPath(partialDiscardFilePath);
    } else {
      this.wholeFileHistory.clearHistory();
    }
  }

  updateHistory(history) {
    this.partialFileHistory.setHistory(history.partialFileHistory || {});
    this.wholeFileHistory.setHistory(history.wholeFileHistory || []);
  }

  createHistoryBlob() {
    var _this = this;

    return _asyncToGenerator(function* () {
      const histories = {
        wholeFileHistory: _this.wholeFileHistory.getHistory(),
        partialFileHistory: _this.partialFileHistory.getHistory()
      };
      const historySha = yield _this.createBlob({ stdin: JSON.stringify(histories) });
      return historySha;
    })();
  }

  storeBeforeAndAfterBlobs(filePaths, isSafe, destructiveAction) {
    var _this2 = this;

    let partialDiscardFilePath = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;
    return _asyncToGenerator(function* () {
      if (partialDiscardFilePath) {
        return yield _this2.storeBlobsForPartialFileHistory(partialDiscardFilePath, isSafe, destructiveAction);
      } else {
        return yield _this2.storeBlobsForWholeFileHistory(filePaths, isSafe, destructiveAction);
      }
    })();
  }

  storeBlobsForPartialFileHistory(filePath, isSafe, destructiveAction) {
    var _this3 = this;

    return _asyncToGenerator(function* () {
      const beforeSha = yield _this3.createBlob({ filePath });
      const isNotSafe = !(yield isSafe());
      if (isNotSafe) {
        return null;
      }
      yield destructiveAction();
      const afterSha = yield _this3.createBlob({ filePath });
      const snapshots = { beforeSha, afterSha };
      _this3.partialFileHistory.addHistory(filePath, snapshots);
      return snapshots;
    })();
  }

  storeBlobsForWholeFileHistory(filePaths, isSafe, destructiveAction) {
    var _this4 = this;

    return _asyncToGenerator(function* () {
      const snapshotsByPath = {};
      const beforePromises = filePaths.map((() => {
        var _ref2 = _asyncToGenerator(function* (filePath) {
          snapshotsByPath[filePath] = { beforeSha: yield _this4.createBlob({ filePath }) };
        });

        return function (_x8) {
          return _ref2.apply(this, arguments);
        };
      })());
      yield Promise.all(beforePromises);
      const isNotSafe = !(yield isSafe());
      if (isNotSafe) {
        return null;
      }
      yield destructiveAction();
      const afterPromises = filePaths.map((() => {
        var _ref3 = _asyncToGenerator(function* (filePath) {
          snapshotsByPath[filePath].afterSha = yield _this4.createBlob({ filePath });
        });

        return function (_x9) {
          return _ref3.apply(this, arguments);
        };
      })());
      yield Promise.all(afterPromises);
      _this4.wholeFileHistory.addHistory(snapshotsByPath);
      return snapshotsByPath;
    })();
  }

  restoreLastDiscardInTempFiles(isSafe) {
    var _this5 = this;

    let partialDiscardFilePath = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
    return _asyncToGenerator(function* () {
      let lastDiscardSnapshots = _this5.getLastSnapshots(partialDiscardFilePath);
      if (partialDiscardFilePath) {
        lastDiscardSnapshots = lastDiscardSnapshots ? [lastDiscardSnapshots] : [];
      }
      const tempFolderPaths = yield _this5.expandBlobsToFilesInTempFolder(lastDiscardSnapshots);
      if (!isSafe()) {
        return [];
      }
      return yield _this5.mergeFiles(tempFolderPaths);
    })();
  }

  expandBlobsToFilesInTempFolder(snapshots) {
    var _this6 = this;

    return _asyncToGenerator(function* () {
      const tempFolderPath = yield (0, _helpers.getTempDir)('github-discard-history-');
      const pathPromises = snapshots.map((() => {
        var _ref4 = _asyncToGenerator(function* (_ref5) {
          let filePath = _ref5.filePath,
              beforeSha = _ref5.beforeSha,
              afterSha = _ref5.afterSha;

          const dir = _path2.default.dirname(_path2.default.join(tempFolderPath, filePath));
          yield (0, _mkdirp2.default)(dir);
          const theirsPath = !beforeSha ? null : yield _this6.expandBlobToFile(_path2.default.join(tempFolderPath, `${filePath}-before-discard`), beforeSha);
          const commonBasePath = !afterSha ? null : yield _this6.expandBlobToFile(_path2.default.join(tempFolderPath, `${filePath}-after-discard`), afterSha);
          const resultPath = _path2.default.join(tempFolderPath, `~${_path2.default.basename(filePath)}-merge-result`);
          return { filePath, commonBasePath, theirsPath, resultPath, theirsSha: beforeSha, commonBaseSha: afterSha };
        });

        return function (_x11) {
          return _ref4.apply(this, arguments);
        };
      })());
      return yield Promise.all(pathPromises);
    })();
  }

  mergeFiles(filePaths) {
    var _this7 = this;

    return _asyncToGenerator(function* () {
      const mergeFilePromises = filePaths.map((() => {
        var _ref6 = _asyncToGenerator(function* (filePathInfo, i) {
          const filePath = filePathInfo.filePath,
                commonBasePath = filePathInfo.commonBasePath,
                theirsPath = filePathInfo.theirsPath,
                resultPath = filePathInfo.resultPath,
                theirsSha = filePathInfo.theirsSha,
                commonBaseSha = filePathInfo.commonBaseSha;

          const currentSha = yield _this7.createBlob({ filePath });
          let mergeResult;
          if (theirsPath && commonBasePath) {
            mergeResult = yield _this7.mergeFile(filePath, commonBasePath, theirsPath, resultPath);
          } else if (!theirsPath && commonBasePath) {
            // deleted file
            const oursSha = yield _this7.createBlob({ filePath });
            if (oursSha === commonBaseSha) {
              // no changes since discard, mark file to be deleted
              mergeResult = { filePath, resultPath: null, deleted: true, conflict: false };
            } else {
              // changes since discard result in conflict
              yield (0, _helpers.copyFile)(_path2.default.join(_this7.workdirPath, filePath), resultPath);
              mergeResult = { filePath, resultPath, conflict: true };
            }
          } else if (theirsPath && !commonBasePath) {
            // added file
            const fileDoesExist = yield (0, _helpers.fileExists)(_path2.default.join(_this7.workdirPath, filePath));
            if (!fileDoesExist) {
              yield (0, _helpers.copyFile)(theirsPath, resultPath);
              mergeResult = { filePath, resultPath, conflict: false };
            } else {
              yield emptyFilePromise;
              mergeResult = yield _this7.mergeFile(filePath, emptyFilePath, theirsPath, resultPath);
            }
          } else {
            throw new Error('One of the following must be defined - theirsPath:' + `${theirsPath} or commonBasePath: ${commonBasePath}`);
          }
          return _extends({}, mergeResult, { theirsSha, commonBaseSha, currentSha });
        });

        return function (_x12, _x13) {
          return _ref6.apply(this, arguments);
        };
      })());
      return yield Promise.all(mergeFilePromises);
    })();
  }
};
exports.default = DiscardHistory;
