// github/lib/models/hunk-line.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _class, _temp;

let HunkLine = (_temp = _class = class HunkLine {

  constructor(text, status, oldLineNumber, newLineNumber, diffLineNumber) {
    this.text = text;
    this.status = status;
    this.oldLineNumber = oldLineNumber;
    this.newLineNumber = newLineNumber;
    this.diffLineNumber = diffLineNumber;
  }

  copy() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    let text = _ref.text,
        status = _ref.status,
        oldLineNumber = _ref.oldLineNumber,
        newLineNumber = _ref.newLineNumber;

    return new HunkLine(text || this.getText(), status || this.getStatus(), oldLineNumber || this.getOldLineNumber(), newLineNumber || this.getNewLineNumber());
  }

  getText() {
    return this.text;
  }

  getOldLineNumber() {
    return this.oldLineNumber;
  }

  getNewLineNumber() {
    return this.newLineNumber;
  }

  getStatus() {
    return this.status;
  }

  isChanged() {
    return this.getStatus() === 'added' || this.getStatus() === 'deleted';
  }

  getOrigin() {
    switch (this.getStatus()) {
      case 'added':
        return '+';
      case 'deleted':
        return '-';
      case 'unchanged':
        return ' ';
      case 'nonewline':
        return '\\';
      default:
        return '';
    }
  }

  invert() {
    let invertedStatus;
    switch (this.getStatus()) {
      case 'added':
        invertedStatus = 'deleted';
        break;
      case 'deleted':
        invertedStatus = 'added';
        break;
      case 'unchanged':
        invertedStatus = 'unchanged';
        break;
      case 'nonewline':
        invertedStatus = 'nonewline';
        break;
    }

    return new HunkLine(this.text, invertedStatus, this.newLineNumber, this.oldLineNumber);
  }

  toString() {
    return this.getOrigin() + (this.getStatus() === 'nonewline' ? ' ' : '') + this.getText();
  }
}, _class.statusMap = {
  '+': 'added',
  '-': 'deleted',
  ' ': 'unchanged',
  '\\': 'nonewline'
}, _temp);
exports.default = HunkLine;
