// github/lib/models/model-observer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

let ModelObserver = class ModelObserver {
  constructor(_ref) {
    let fetchData = _ref.fetchData,
        didUpdate = _ref.didUpdate;

    this.fetchData = fetchData || (() => {});
    this.didUpdate = didUpdate || (() => {});
    this.activeModel = null;
    this.activeModelData = null;
    this.activeModelUpdateSubscription = null;
    this.inProgress = false;
    this.pending = false;
  }

  setActiveModel(model) {
    if (model !== this.activeModel) {
      if (this.activeModelUpdateSubscription) {
        this.activeModelUpdateSubscription.dispose();
        this.activeModelUpdateSubscription = null;
      }
      this.activeModel = model;
      this.activeModelData = null;
      this.inProgress = false;
      this.pending = false;
      this.didUpdate(model);
      if (model) {
        this.activeModelUpdateSubscription = model.onDidUpdate(() => this.refreshModelData(model));
        return this.refreshModelData(model);
      }
    }
    return null;
  }

  refreshModelData() {
    let model = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.activeModel;

    if (this.inProgress) {
      this.pending = true;
      return null;
    }
    this.lastModelDataRefreshPromise = this._refreshModelData(model);
    return this.lastModelDataRefreshPromise;
  }

  _refreshModelData(model) {
    var _this = this;

    return _asyncToGenerator(function* () {
      try {
        _this.inProgress = true;
        const fetchDataPromise = _this.fetchData(model);
        _this.lastFetchDataPromise = fetchDataPromise;
        const modelData = yield fetchDataPromise;
        // Since we re-fetch immediately when the model changes,
        // we need to ensure a fetch for an old active model
        // does not trample the newer fetch for the newer active model.
        if (fetchDataPromise === _this.lastFetchDataPromise) {
          _this.activeModel = model;
          _this.activeModelData = modelData;
          _this.didUpdate(model);
        }
      } finally {
        _this.inProgress = false;
        if (_this.pending) {
          _this.pending = false;
          _this.refreshModelData();
        }
      }
    })();
  }

  getActiveModel() {
    return this.activeModel;
  }

  getActiveModelData() {
    return this.activeModelData;
  }

  getLastModelDataRefreshPromise() {
    return this.lastModelDataRefreshPromise;
  }

  destroy() {
    if (this.activeModelUpdateSubscription) {
      this.activeModelUpdateSubscription.dispose();
    }
  }
};
exports.default = ModelObserver;
