// github/lib/models/remote.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
let Remote = class Remote {
  constructor(name, url) {
    this.name = name;
    this.url = url;

    var _githubInfoFromRemote = githubInfoFromRemote(url);

    const isGithubRepo = _githubInfoFromRemote.isGithubRepo,
          owner = _githubInfoFromRemote.owner,
          repo = _githubInfoFromRemote.repo;

    this.githubRepo = isGithubRepo;
    this.owner = owner;
    this.repo = repo;
  }

  getName() {
    return this.name;
  }

  getUrl() {
    return this.url;
  }

  isGithubRepo() {
    return this.githubRepo;
  }

  getOwner() {
    return this.owner;
  }

  getRepo() {
    return this.repo;
  }

  getNameOr(fallback) {
    return this.getName();
  }

  isPresent() {
    return true;
  }
};
exports.default = Remote;


function githubInfoFromRemote(remoteUrl) {
  if (!remoteUrl) {
    return {
      isGithubRepo: false,
      owner: null,
      repo: null
    };
  }

  //             proto       login   domain         owner    repo
  const regex = /(?:.+:\/\/)?(?:.+@)?github\.com[:/]([^/]+)\/(.+)/;
  const match = remoteUrl.match(regex);
  if (match) {
    return {
      isGithubRepo: true,
      owner: match[1],
      repo: match[2].replace(/\.git$/, '')
    };
  } else {
    return {
      isGithubRepo: false,
      owner: null,
      repo: null
    };
  }
}

const nullRemote = exports.nullRemote = {
  getName() {
    return '';
  },

  getUrl() {
    return '';
  },

  isGithubRepo() {
    return false;
  },

  getOwner() {
    return null;
  },

  getRepo() {
    return null;
  },

  getNameOr(fallback) {
    return fallback;
  },

  isPresent() {
    return false;
  }
};
