// github/lib/models/workdir-cache.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _helpers = require('../helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

/**
 * Locate the nearest git working directory above a given starting point, caching results.
 */
let WorkdirCache = class WorkdirCache {
  constructor() {
    let maxSize = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1000;

    this.maxSize = maxSize;
    this.known = new Map();
  }

  find(startDir) {
    var _this = this;

    return _asyncToGenerator(function* () {
      try {
        const resolvedDir = yield _this.resolvePath(startDir);
        const cached = _this.known.get(resolvedDir);
        if (cached !== undefined) {
          return cached;
        }

        const workDir = yield _this.walkToRoot(resolvedDir);

        if (_this.known.size >= _this.maxSize) {
          _this.known.clear();
        }
        _this.known.set(resolvedDir, workDir);
        return workDir;
      } catch (e) {
        if (e.code === 'ENOENT') {
          return null;
        }

        throw e;
      }
    })();
  }

  invalidate(baseDir) {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      const resolvedBase = yield _this2.resolvePath(baseDir);
      for (const cachedPath of _this2.known.keys()) {
        if (cachedPath.startsWith(resolvedBase)) {
          _this2.known.delete(cachedPath);
        }
      }
    })();
  }

  resolvePath(unresolvedPath) {
    return new Promise((resolve, reject) => {
      _fs2.default.realpath(unresolvedPath, (err, resolved) => err ? reject(err) : resolve(resolved));
    });
  }

  walkToRoot(initialDir) {
    return new Promise((resolve, reject) => {
      let currentDir = initialDir;

      const check = () => {
        const dotGit = _path2.default.join(currentDir, '.git');
        _fs2.default.stat(dotGit, (() => {
          var _ref = _asyncToGenerator(function* (statError, stat) {
            if (statError) {
              if (statError.code === 'ENOENT' || statError.code === 'ENOTDIR') {
                // File not found. This is not the directory we're looking for. Continue walking.
                return walk();
              }

              return reject(statError);
            }

            if (!stat.isDirectory()) {
              const contents = yield (0, _helpers.readFile)(dotGit, 'utf8');
              if (contents.startsWith('gitdir: ')) {
                return resolve(currentDir);
              } else {
                return walk();
              }
            }

            // .git directory found! Mission accomplished.
            return resolve(currentDir);
          });

          return function (_x2, _x3) {
            return _ref.apply(this, arguments);
          };
        })());
      };

      const walk = () => {
        const parentDir = _path2.default.resolve(currentDir, '..');
        if (parentDir === currentDir) {
          // Root directory. Traversal done, no working directory found.
          resolve(null);
          return;
        }

        currentDir = parentDir;
        check();
      };

      check();
    });
  }
};
exports.default = WorkdirCache;
