// about/lib/update-manager.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

/** @babel */

var _atom = require('atom');

var Unsupported = 'unsupported';
var Idle = 'idle';
var CheckingForUpdate = 'checking';
var DownloadingUpdate = 'downloading';
var UpdateAvailableToInstall = 'update-available';
var UpToDate = 'no-update-available';
var ErrorState = 'error';

var UpdateManager = (function () {
  function UpdateManager() {
    _classCallCheck(this, UpdateManager);

    this.emitter = new _atom.Emitter();
    this.currentVersion = atom.getVersion();
    this.availableVersion = atom.getVersion();
    this.resetState();
    this.listenForAtomEvents();
  }

  _createClass(UpdateManager, [{
    key: 'listenForAtomEvents',
    value: function listenForAtomEvents() {
      var _this = this;

      this.subscriptions = new _atom.CompositeDisposable();

      this.subscriptions.add(atom.autoUpdater.onDidBeginCheckingForUpdate(function () {
        _this.setState(CheckingForUpdate);
      }), atom.autoUpdater.onDidBeginDownloadingUpdate(function () {
        _this.setState(DownloadingUpdate);
      }), atom.autoUpdater.onDidCompleteDownloadingUpdate(function (_ref) {
        var releaseVersion = _ref.releaseVersion;

        _this.setAvailableVersion(releaseVersion);
      }), atom.autoUpdater.onUpdateNotAvailable(function () {
        _this.setState(UpToDate);
      }), atom.autoUpdater.onUpdateError(function () {
        _this.setState(ErrorState);
      }), atom.config.observe('core.automaticallyUpdate', function (value) {
        _this.autoUpdatesEnabled = value;
        _this.emitDidChange();
      }));

      // TODO: When https://github.com/atom/electron/issues/4587 is closed we can add this support.
      // atom.autoUpdater.onUpdateAvailable =>
      //   @find('.about-updates-item').removeClass('is-shown')
      //   @updateAvailable.addClass('is-shown')
    }
  }, {
    key: 'dispose',
    value: function dispose() {
      this.subscriptions.dispose();
    }
  }, {
    key: 'onDidChange',
    value: function onDidChange(callback) {
      return this.emitter.on('did-change', callback);
    }
  }, {
    key: 'emitDidChange',
    value: function emitDidChange() {
      this.emitter.emit('did-change');
    }
  }, {
    key: 'getAutoUpdatesEnabled',
    value: function getAutoUpdatesEnabled() {
      return this.autoUpdatesEnabled && this.state !== UpdateManager.State.Unsupported;
    }
  }, {
    key: 'setAutoUpdatesEnabled',
    value: function setAutoUpdatesEnabled(enabled) {
      return atom.config.set('core.automaticallyUpdate', enabled);
    }
  }, {
    key: 'getErrorMessage',
    value: function getErrorMessage() {
      return atom.autoUpdater.getErrorMessage();
    }
  }, {
    key: 'getState',
    value: function getState() {
      return this.state;
    }
  }, {
    key: 'setState',
    value: function setState(state) {
      this.state = state;
      this.emitDidChange();
    }
  }, {
    key: 'resetState',
    value: function resetState() {
      this.state = atom.autoUpdater.platformSupportsUpdates() ? atom.autoUpdater.getState() : Unsupported;
      this.emitDidChange();
    }
  }, {
    key: 'getAvailableVersion',
    value: function getAvailableVersion() {
      return this.availableVersion;
    }
  }, {
    key: 'setAvailableVersion',
    value: function setAvailableVersion(version) {
      this.availableVersion = version;

      if (this.availableVersion !== this.currentVersion) {
        this.state = UpdateAvailableToInstall;
      } else {
        this.state = UpToDate;
      }

      this.emitDidChange();
    }
  }, {
    key: 'checkForUpdate',
    value: function checkForUpdate() {
      atom.autoUpdater.checkForUpdate();
    }
  }, {
    key: 'restartAndInstallUpdate',
    value: function restartAndInstallUpdate() {
      atom.autoUpdater.restartAndInstallUpdate();
    }
  }, {
    key: 'getReleaseNotesURLForCurrentVersion',
    value: function getReleaseNotesURLForCurrentVersion() {
      return this.getReleaseNotesURLForVersion(this.currentVersion);
    }
  }, {
    key: 'getReleaseNotesURLForAvailableVersion',
    value: function getReleaseNotesURLForAvailableVersion() {
      return this.getReleaseNotesURLForVersion(this.availableVersion);
    }
  }, {
    key: 'getReleaseNotesURLForVersion',
    value: function getReleaseNotesURLForVersion(appVersion) {
      // Dev versions will not have a releases page
      if (appVersion.indexOf('dev') > -1) {
        return 'https://atom.io/releases';
      }

      if (!appVersion.startsWith('v')) {
        appVersion = 'v' + appVersion;
      }
      return 'https://github.com/atom/atom/releases/tag/' + appVersion;
    }
  }]);

  return UpdateManager;
})();

exports['default'] = UpdateManager;

UpdateManager.State = {
  Unsupported: Unsupported,
  Idle: Idle,
  CheckingForUpdate: CheckingForUpdate,
  DownloadingUpdate: DownloadingUpdate,
  UpdateAvailableToInstall: UpdateAvailableToInstall,
  UpToDate: UpToDate,
  Error: ErrorState
};
module.exports = exports['default'];
