// bracket-matcher/lib/match-manager.js
(function() {
  var CompositeDisposable, MatchManager, _,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  _ = require('underscore-plus');

  CompositeDisposable = require('atom').CompositeDisposable;

  module.exports = MatchManager = (function() {
    MatchManager.prototype.appendPair = function(pairList, arg) {
      var itemLeft, itemRight, newPair;
      itemLeft = arg[0], itemRight = arg[1];
      newPair = {};
      newPair[itemLeft] = itemRight;
      return pairList = _.extend(pairList, newPair);
    };

    MatchManager.prototype.processAutoPairs = function(autocompletePairs, pairedList, dataFun) {
      var autocompletePair, i, len, pairArray, results;
      if (autocompletePairs.length) {
        results = [];
        for (i = 0, len = autocompletePairs.length; i < len; i++) {
          autocompletePair = autocompletePairs[i];
          pairArray = autocompletePair.split('');
          results.push(this.appendPair(pairedList, dataFun(pairArray)));
        }
        return results;
      }
    };

    MatchManager.prototype.updateConfig = function() {
      var endPair, ref, results, startPair;
      this.pairedCharacters = {};
      this.pairedCharactersInverse = {};
      this.pairRegexes = {};
      this.pairsWithExtraNewline = {};
      this.processAutoPairs(this.getScopedSetting('bracket-matcher.autocompleteCharacters'), this.pairedCharacters, (function(x) {
        return [x[0], x[1]];
      }));
      this.processAutoPairs(this.getScopedSetting('bracket-matcher.autocompleteCharacters'), this.pairedCharactersInverse, (function(x) {
        return [x[1], x[0]];
      }));
      this.processAutoPairs(this.getScopedSetting('bracket-matcher.pairsWithExtraNewline'), this.pairsWithExtraNewline, (function(x) {
        return [x[0], x[1]];
      }));
      ref = this.pairedCharacters;
      results = [];
      for (startPair in ref) {
        endPair = ref[startPair];
        results.push(this.pairRegexes[startPair] = new RegExp("[" + (_.escapeRegExp(startPair + endPair)) + "]", 'g'));
      }
      return results;
    };

    MatchManager.prototype.getScopedSetting = function(key) {
      return atom.config.get(key, {
        scope: this.editor.getRootScopeDescriptor()
      });
    };

    function MatchManager(editor, editorElement) {
      this.editor = editor;
      this.destroy = bind(this.destroy, this);
      this.subscriptions = new CompositeDisposable;
      this.updateConfig();
      this.subscriptions.add(atom.config.observe('bracket-matcher.autocompleteCharacters', {
        scope: this.editor.getRootScopeDescriptor()
      }, (function(_this) {
        return function(newConfig) {
          return _this.updateConfig();
        };
      })(this)));
      this.subscriptions.add(atom.config.observe('bracket-matcher.pairsWithExtraNewline', {
        scope: this.editor.getRootScopeDescriptor()
      }, (function(_this) {
        return function(newConfig) {
          return _this.updateConfig();
        };
      })(this)));
    }

    MatchManager.prototype.destroy = function() {
      return this.subscriptions.dispose();
    };

    return MatchManager;

  })();

}).call(this);
