// github/lib/async-queue.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

let Task = class Task {
  constructor(fn) {
    let parallel = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;

    this.fn = fn;
    this.parallel = parallel;
    this.promise = new Promise((resolve, reject) => {
      this.resolve = resolve;
      this.reject = reject;
    });
  }

  execute() {
    var _this = this;

    return _asyncToGenerator(function* () {
      try {
        const value = yield _this.fn.call(undefined);
        _this.resolve(value);
      } catch (err) {
        _this.reject(err);
      }
    })();
  }

  runsInParallel() {
    return this.parallel;
  }

  runsInSerial() {
    return !this.parallel;
  }

  getPromise() {
    return this.promise;
  }
};
let AsyncQueue = class AsyncQueue {
  constructor() {
    let options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    this.parallelism = options.parallelism || 1;
    this.nonParallelizableOperation = false;
    this.tasksInProgress = 0;
    this.queue = [];
  }

  push(fn) {
    var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : { parallel: true };

    let parallel = _ref.parallel;

    const task = new Task(fn, parallel);
    this.queue.push(task);
    this.processQueue();
    return task.getPromise();
  }

  processQueue() {
    if (!this.queue.length || this.nonParallelizableOperation || this.disposed) {
      return;
    }

    const task = this.queue[0];
    const canRunParallelOp = task.runsInParallel() && this.tasksInProgress < this.parallelism;
    const canRunSerialOp = task.runsInSerial() && this.tasksInProgress === 0;
    if (canRunSerialOp || canRunParallelOp) {
      this.processTask(task, task.runsInParallel());
      this.queue.shift();
      this.processQueue();
    }
  }

  processTask(task, runsInParallel) {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      if (_this2.disposed) {
        return;
      }

      _this2.tasksInProgress++;
      if (!runsInParallel) {
        _this2.nonParallelizableOperation = true;
      }

      try {
        yield task.execute();
      } finally {
        _this2.tasksInProgress--;
        _this2.nonParallelizableOperation = false;
        _this2.processQueue();
      }
    })();
  }

  dispose() {
    this.queue = [];
    this.disposed = true;
  }
};
exports.default = AsyncQueue;
