// github/lib/context-menu-interceptor.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let ContextMenuInterceptor = (_temp = _class = class ContextMenuInterceptor extends _react2.default.Component {

  static handle(event) {
    for (const _ref of ContextMenuInterceptor.registration) {
      var _ref2 = _slicedToArray(_ref, 2);

      const element = _ref2[0];
      const callback = _ref2[1];

      if (element.contains(event.target)) {
        callback(event);
      }
    }
  }

  static dispose() {
    document.removeEventListener('contextmenu', contextMenuHandler, { capture: true });
  }

  componentDidMount() {
    // Helpfully, addEventListener dedupes listeners for us.
    document.addEventListener('contextmenu', contextMenuHandler, { capture: true });
    ContextMenuInterceptor.registration.set(this.element, this.props.onWillShowContextMenu);
  }

  render() {
    return _react2.default.createElement(
      'div',
      { ref: e => {
          this.element = e;
        } },
      this.props.children
    );
  }

  componentWillUnmount() {
    ContextMenuInterceptor.registration.delete(this.element);
  }
}, _class.propTypes = {
  onWillShowContextMenu: _propTypes2.default.func.isRequired,
  children: _propTypes2.default.element.isRequired
}, _class.registration = new Map(), _temp);
exports.default = ContextMenuInterceptor;


function contextMenuHandler(event) {
  ContextMenuInterceptor.handle(event);
}
