// github/lib/deferred-callback-queue.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _desc, _value, _class;

var _electron = require('electron');

var _eventKit = require('event-kit');

var _coreDecorators = require('core-decorators');

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

let DeferredCallbackQueue = (_class = class DeferredCallbackQueue {
  constructor(wait, callback) {
    let options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

    this.wait = wait;
    const currentWindow = _electron.remote.getCurrentWindow();

    const onDidFocus = options.onDidFocus || function (cb) {
      currentWindow.on('focus', cb);
      return new _eventKit.Disposable(() => {
        currentWindow.removeListener('focus', cb);
      });
    };
    const onDidBlur = options.onDidBlur || function (cb) {
      if (atom.inSpecMode()) {
        return new _eventKit.Disposable();
      }
      currentWindow.on('blur', cb);
      return new _eventKit.Disposable(() => {
        currentWindow.removeListener('blur', cb);
      });
    };

    this.callback = callback;
    this.subscriptions = new _eventKit.CompositeDisposable();
    this.items = new Set();

    this.paused = !currentWindow.isFocused() && !atom.inSpecMode();
    this.subscriptions.add(onDidFocus(this.resume));
    this.subscriptions.add(onDidBlur(this.pause));
  }

  pause() {
    if (this.paused) {
      return;
    }
    this.paused = true;
  }

  resume() {
    if (!this.paused) {
      return;
    }
    this.paused = false;
    if (this.timer) {
      clearTimeout(this.timer);
    }
    this.flush();
  }

  resetTimer() {
    if (this.timer) {
      clearTimeout(this.timer);
    }
    this.timer = setTimeout(this.flush, this.wait);
  }

  flush() {
    delete this.timer;
    if (this.items.size) {
      this.callback([...this.items]);
      this.items.clear();
    }
  }

  push() {
    for (var _len = arguments.length, items = Array(_len), _key = 0; _key < _len; _key++) {
      items[_key] = arguments[_key];
    }

    if (this.paused) {
      items.forEach(item => this.items.add(item));
      this.resetTimer();
    } else {
      this.callback(items);
    }
  }

  destroy() {
    this.subscriptions.dispose();
  }
}, (_applyDecoratedDescriptor(_class.prototype, 'pause', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'pause'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'resume', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'resume'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'flush', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'flush'), _class.prototype)), _class);
exports.default = DeferredCallbackQueue;
