// github/lib/prop-types.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommitPropType = exports.BranchPropType = exports.RemotePropType = exports.DOMNodePropType = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const DOMNodePropType = exports.DOMNodePropType = (props, propName, componentName) => {
  if (props[propName] instanceof HTMLElement) {
    return null;
  } else {
    return new Error(`Invalid prop '${propName}' supplied to '${componentName}'. Value is not DOM element.`);
  }
};

const RemotePropType = exports.RemotePropType = _propTypes2.default.shape({
  getName: _propTypes2.default.func.isRequired,
  getUrl: _propTypes2.default.func.isRequired,
  isGithubRepo: _propTypes2.default.func.isRequired,
  getOwner: _propTypes2.default.func.isRequired,
  getRepo: _propTypes2.default.func.isRequired
});

const BranchPropType = exports.BranchPropType = _propTypes2.default.shape({
  getName: _propTypes2.default.func.isRequired,
  isDetached: _propTypes2.default.func.isRequired,
  isPresent: _propTypes2.default.func.isRequired
});

const CommitPropType = exports.CommitPropType = _propTypes2.default.shape({
  getSha: _propTypes2.default.func.isRequired,
  getMessage: _propTypes2.default.func.isRequired,
  isUnbornRef: _propTypes2.default.func.isRequired,
  isPresent: _propTypes2.default.func.isRequired
});
