// image-view/lib/image-editor.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

/** @babel */

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fsPlus = require('fs-plus');

var _fsPlus2 = _interopRequireDefault(_fsPlus);

var _atom = require('atom');

var _imageEditorView = require('./image-editor-view');

var _imageEditorView2 = _interopRequireDefault(_imageEditorView);

// Editor model for an image file

var ImageEditor = (function () {
  _createClass(ImageEditor, null, [{
    key: 'deserialize',
    value: function deserialize(_ref) {
      var filePath = _ref.filePath;

      if (_fsPlus2['default'].isFileSync(filePath)) {
        return new ImageEditor(filePath);
      } else {
        console.warn('Could not deserialize image editor for path \'' + filePath + '\' because that file no longer exists');
      }
    }
  }]);

  function ImageEditor(filePath) {
    _classCallCheck(this, ImageEditor);

    this.file = new _atom.File(filePath);
    this.subscriptions = new _atom.CompositeDisposable();
    this.emitter = new _atom.Emitter();
  }

  _createClass(ImageEditor, [{
    key: 'serialize',
    value: function serialize() {
      return { filePath: this.getPath(), deserializer: this.constructor.name };
    }
  }, {
    key: 'terminatePendingState',
    value: function terminatePendingState() {
      if (this.isEqual(atom.workspace.getActivePane().getPendingItem())) {
        this.emitter.emit('did-terminate-pending-state');
      }
    }
  }, {
    key: 'onDidTerminatePendingState',
    value: function onDidTerminatePendingState(callback) {
      return this.emitter.on('did-terminate-pending-state', callback);
    }

    // Register a callback for when the image file changes
  }, {
    key: 'onDidChange',
    value: function onDidChange(callback) {
      var changeSubscription = this.file.onDidChange(callback);
      this.subscriptions.add(changeSubscription);
      return changeSubscription;
    }

    // Register a callback for whne the image's title changes
  }, {
    key: 'onDidChangeTitle',
    value: function onDidChangeTitle(callback) {
      var renameSubscription = this.file.onDidRename(callback);
      this.subscriptions.add(renameSubscription);
      return renameSubscription;
    }
  }, {
    key: 'destroy',
    value: function destroy() {
      this.subscriptions.dispose();
      if (this.view) {
        this.view.destroy();
      }
    }

    // Retrieves the filename of the open file.
    //
    // This is `'untitled'` if the file is new and not saved to the disk.
    //
    // Returns a {String}.
  }, {
    key: 'getTitle',
    value: function getTitle() {
      var filePath = this.getPath();
      if (filePath) {
        return _path2['default'].basename(filePath);
      } else {
        return 'untitled';
      }
    }

    // Retrieves the absolute path to the image.
    //
    // Returns a {String} path.
  }, {
    key: 'getPath',
    value: function getPath() {
      return this.file.getPath();
    }

    // Retrieves the URI of the image.
    //
    // Returns a {String}.
  }, {
    key: 'getURI',
    value: function getURI() {
      return this.getPath();
    }

    // Retrieves the encoded URI of the image.
    //
    // Returns a {String}.
  }, {
    key: 'getEncodedURI',
    value: function getEncodedURI() {
      return 'file://' + encodeURI(this.getPath().replace(/\\/g, '/')).replace(/#/g, '%23').replace(/\?/g, '%3F');
    }

    // Compares two {ImageEditor}s to determine equality.
    //
    // Equality is based on the condition that the two URIs are the same.
    //
    // Returns a {Boolean}.
  }, {
    key: 'isEqual',
    value: function isEqual(other) {
      return other instanceof ImageEditor && this.getURI() === other.getURI();
    }

    // Essential: Invoke the given callback when the editor is destroyed.
    //
    // * `callback` {Function} to be called when the editor is destroyed.
    //
    // Returns a {Disposable} on which `.dispose()` can be called to unsubscribe.
  }, {
    key: 'onDidDestroy',
    value: function onDidDestroy(callback) {
      return this.emitter.on('did-destroy', callback);
    }
  }, {
    key: 'element',
    get: function get() {
      return this.view.element;
    }
  }, {
    key: 'view',
    get: function get() {
      if (!this.editorView) {
        this.editorView = new _imageEditorView2['default'](this);
      }
      return this.editorView;
    }
  }]);

  return ImageEditor;
})();

exports['default'] = ImageEditor;
module.exports = exports['default'];
