// image-view/lib/image-editor-status-view.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

/** @babel */

var _atom = require('atom');

var _imageEditor = require('./image-editor');

var _imageEditor2 = _interopRequireDefault(_imageEditor);

var _bytes = require('bytes');

var _bytes2 = _interopRequireDefault(_bytes);

var ImageEditorStatusView = (function () {
  function ImageEditorStatusView(statusBar) {
    var _this = this;

    _classCallCheck(this, ImageEditorStatusView);

    this.statusBar = statusBar;
    this.disposables = new _atom.CompositeDisposable();

    this.element = document.createElement('div');
    this.element.classList.add('status-image', 'inline-block');

    this.imageSizeStatus = document.createElement('span');
    this.imageSizeStatus.classList.add('image-size');
    this.element.appendChild(this.imageSizeStatus);

    this.attach();

    this.disposables.add(atom.workspace.onDidChangeActivePaneItem(function () {
      _this.updateImageSize();
    }));
  }

  _createClass(ImageEditorStatusView, [{
    key: 'attach',
    value: function attach() {
      this.statusBarTile = this.statusBar.addLeftTile({ item: this });
      this.updateImageSize();
    }
  }, {
    key: 'destroy',
    value: function destroy() {
      this.statusBarTile.destroy();
      this.disposables.dispose();
    }
  }, {
    key: 'getImageSize',
    value: function getImageSize(_ref) {
      var originalHeight = _ref.originalHeight;
      var originalWidth = _ref.originalWidth;
      var imageSize = _ref.imageSize;

      this.imageSizeStatus.textContent = originalWidth + 'x' + originalHeight + ' ' + (0, _bytes2['default'])(imageSize);
      this.imageSizeStatus.style.display = '';
    }
  }, {
    key: 'updateImageSize',
    value: function updateImageSize() {
      var _this2 = this;

      if (this.imageLoadDisposable) {
        this.imageLoadDisposable.dispose();
      }

      var editor = atom.workspace.getActivePaneItem();
      if (editor instanceof _imageEditor2['default']) {
        this.editorView = editor.view;
        if (this.editorView.loaded) {
          this.getImageSize(this.editorView);
        }

        this.imageLoadDisposable = this.editorView.onDidLoad(function () {
          if (editor === atom.workspace.getActivePaneItem()) {
            _this2.getImageSize(_this2.editorView);
          }
        });
      } else {
        this.imageSizeStatus.style.display = 'none';
      }
    }
  }]);

  return ImageEditorStatusView;
})();

exports['default'] = ImageEditorStatusView;
module.exports = exports['default'];
