// dev-live-reload/lib/ui-watcher.js
(function() {
  var BaseThemeWatcher, PackageWatcher, UIWatcher, _,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  _ = require('underscore-plus');

  BaseThemeWatcher = require('./base-theme-watcher');

  PackageWatcher = require('./package-watcher');

  module.exports = UIWatcher = (function() {
    function UIWatcher() {
      this.reloadAll = bind(this.reloadAll, this);
      this.watchers = [];
      this.baseTheme = this.createWatcher(BaseThemeWatcher);
      this.watchPackages();
    }

    UIWatcher.prototype.watchPackages = function() {
      var i, j, len, len1, pack, ref, ref1, theme;
      this.watchedThemes = {};
      this.watchedPackages = {};
      ref = atom.themes.getActiveThemes();
      for (i = 0, len = ref.length; i < len; i++) {
        theme = ref[i];
        this.watchTheme(theme);
      }
      ref1 = atom.packages.getLoadedPackages();
      for (j = 0, len1 = ref1.length; j < len1; j++) {
        pack = ref1[j];
        this.watchPackage(pack);
      }
      return this.watchForPackageChanges();
    };

    UIWatcher.prototype.watchForPackageChanges = function() {
      return atom.themes.onDidChangeActiveThemes((function(_this) {
        return function() {
          var i, len, name, ref, ref1, theme, watcher;
          ref = _this.watchedThemes;
          for (name in ref) {
            watcher = ref[name];
            watcher.destroy();
          }
          _this.watchedThemes = {};
          ref1 = atom.themes.getActiveThemes();
          for (i = 0, len = ref1.length; i < len; i++) {
            theme = ref1[i];
            _this.watchTheme(theme);
          }
          return null;
        };
      })(this));
    };

    UIWatcher.prototype.watchTheme = function(theme) {
      if (PackageWatcher.supportsPackage(theme, 'theme')) {
        return this.watchedThemes[theme.name] = this.createWatcher(PackageWatcher, theme);
      }
    };

    UIWatcher.prototype.watchPackage = function(pack) {
      if (PackageWatcher.supportsPackage(pack, 'atom')) {
        return this.watchedPackages[pack.name] = this.createWatcher(PackageWatcher, pack);
      }
    };

    UIWatcher.prototype.createWatcher = function(type, object) {
      var watcher;
      watcher = new type(object);
      watcher.onDidChangeGlobals((function(_this) {
        return function() {
          console.log('Global changed, reloading all styles');
          return _this.reloadAll();
        };
      })(this));
      watcher.onDidDestroy((function(_this) {
        return function() {
          return _this.watchers = _.without(_this.watchers, watcher);
        };
      })(this));
      this.watchers.push(watcher);
      return watcher;
    };

    UIWatcher.prototype.reloadAll = function() {
      var i, j, len, len1, pack, ref, ref1, results;
      this.baseTheme.loadAllStylesheets();
      ref = atom.packages.getActivePackages();
      for (i = 0, len = ref.length; i < len; i++) {
        pack = ref[i];
        if (PackageWatcher.supportsPackage(pack, 'atom')) {
          pack.reloadStylesheets();
        }
      }
      ref1 = atom.themes.getActiveThemes();
      results = [];
      for (j = 0, len1 = ref1.length; j < len1; j++) {
        pack = ref1[j];
        if (PackageWatcher.supportsPackage(pack, 'theme')) {
          results.push(pack.reloadStylesheets());
        }
      }
      return results;
    };

    UIWatcher.prototype.destroy = function() {
      var name, ref, ref1, results, watcher;
      this.baseTheme.destroy();
      ref = this.watchedPackages;
      for (name in ref) {
        watcher = ref[name];
        watcher.destroy();
      }
      ref1 = this.watchedThemes;
      results = [];
      for (name in ref1) {
        watcher = ref1[name];
        results.push(watcher.destroy());
      }
      return results;
    };

    return UIWatcher;

  })();

}).call(this);
