// styleguide/lib/code-block.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

/** @babel */

var _highlights = require('highlights');

var _highlights2 = _interopRequireDefault(_highlights);

var CodeBlock = (function () {
  function CodeBlock(props) {
    _classCallCheck(this, CodeBlock);

    this.highlighter = new _highlights2['default']({ registry: atom.grammars, scopePrefix: 'syntax--' });
    this.wrapperElement = document.createElement('div');
    this.update(props);
  }

  _createClass(CodeBlock, [{
    key: 'update',
    value: function update(_ref) {
      var cssClass = _ref.cssClass;
      var grammarScopeName = _ref.grammarScopeName;
      var code = _ref.code;

      this.wrapperElement.innerHTML = this.highlighter.highlightSync({ fileContents: code, scopeName: grammarScopeName });
      this.element = this.wrapperElement.children[0];
      this.element.classList.remove('editor'); // The `editor` class messes things up as `.editor` has absolutely positioned lines
      this.element.classList.add(cssClass);
      var fontFamily = atom.config.get('editor.fontFamily');
      if (fontFamily) {
        this.element.style.fontFamily = fontFamily;
      }
    }
  }]);

  return CodeBlock;
})();

exports['default'] = CodeBlock;
module.exports = exports['default'];
