// styleguide/lib/styleguide-section.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

/** @babel */
/** @jsx etch.dom */

var _etch = require('etch');

var _etch2 = _interopRequireDefault(_etch);

var StyleguideSection = (function () {
  function StyleguideSection(props, children) {
    _classCallCheck(this, StyleguideSection);

    this.collapsed = props.collapsed;
    this.title = props.title;
    this.name = props.name;
    this.children = children;
    _etch2['default'].initialize(this);
    if (props.onDidInitialize) {
      props.onDidInitialize(this);
    }
  }

  _createClass(StyleguideSection, [{
    key: 'render',
    value: function render() {
      var _this = this;

      if (this.loaded) {
        var className = 'bordered';
        if (this.collapsed) {
          className += ' collapsed';
        }
        return _etch2['default'].dom(
          'section',
          { className: className, dataset: { name: this.name } },
          _etch2['default'].dom(
            'h1',
            { className: 'section-heading', onclick: function () {
                return _this.toggle();
              } },
            this.title
          ),
          this.children
        );
      } else {
        return _etch2['default'].dom(
          'section',
          { className: 'bordered collapsed', dataset: { name: this.name } },
          _etch2['default'].dom(
            'h1',
            { className: 'section-heading', onclick: function () {
                return _this.toggle();
              } },
            this.title
          )
        );
      }
    }
  }, {
    key: 'update',
    value: function update(props, children) {
      if (props.title) {
        this.title = props.title;
      }

      if (props.name) {
        this.name = props.name;
      }

      if (children) {
        this.children = children;
      }

      if (props.didExpandOrCollapseSection) {
        this.didExpandOrCollapseSection = props.onDidExpandOrCollapseSection;
      }

      return _etch2['default'].update(this);
    }
  }, {
    key: 'toggle',
    value: function toggle() {
      return this.collapsed ? this.expand() : this.collapse();
    }
  }, {
    key: 'expand',
    value: function expand() {
      this.collapsed = false;
      this.loaded = true;
      return _etch2['default'].update(this);
    }
  }, {
    key: 'collapse',
    value: function collapse() {
      this.collapsed = true;
      return _etch2['default'].update(this);
    }
  }]);

  return StyleguideSection;
})();

exports['default'] = StyleguideSection;
module.exports = exports['default'];
