// github/lib/routes/issueish-info-by-number-route.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _class, _temp;

var _reactRelay = require('react-relay');

var _reactRelay2 = _interopRequireDefault(_reactRelay);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let IssueishInfoByNumberRoute = (_temp = _class = class IssueishInfoByNumberRoute extends _reactRelay2.default.Route {}, _class.routeName = 'pr-info-by-number-route', _class.queries = {
  query: (Component, variables) => function (RQL_0) {
    return {
      children: [].concat.apply([], [_reactRelay2.default.QL.__frag(RQL_0)]),
      fieldName: 'relay',
      kind: 'Query',
      metadata: {},
      name: 'Issueish',
      type: 'Query'
    };
  }(Component.getFragment('query', variables))
}, _class.paramDefinitions = {
  repoOwner: { required: true },
  repoName: { required: true },
  issueishNumber: { required: true }
}, _temp);
exports.default = IssueishInfoByNumberRoute;
