// github/lib/routes/issueish-tooltip-route.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _class, _temp;

var _reactRelay = require('react-relay');

var _reactRelay2 = _interopRequireDefault(_reactRelay);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let IssueishTooltipRoute = (_temp = _class = class IssueishTooltipRoute extends _reactRelay2.default.Route {}, _class.routeName = 'issueish-tooltip-route', _class.queries = {
  resource: Component => function (RQL_0) {
    return {
      calls: [{
        kind: 'Call',
        metadata: {
          type: 'URI!'
        },
        name: 'url',
        value: {
          kind: 'CallVariable',
          callVariableName: 'issueishUrl'
        }
      }],
      children: [].concat.apply([], [{
        fieldName: '__typename',
        kind: 'Field',
        metadata: {
          isGenerated: true,
          isRequisite: true
        },
        type: 'String'
      }, _reactRelay2.default.QL.__frag(RQL_0)]),
      fieldName: 'resource',
      kind: 'Query',
      metadata: {
        isAbstract: true,
        identifyingArgName: 'url',
        identifyingArgType: 'URI!'
      },
      name: 'Issueish',
      type: 'UniformResourceLocatable'
    };
  }(Component.getFragment('resource'))
}, _class.paramDefinitions = {
  issueishUrl: { required: true }
}, _temp);
exports.default = IssueishTooltipRoute;
