// github/lib/routes/user-mention-tooltip-route.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _class, _temp;

var _reactRelay = require('react-relay');

var _reactRelay2 = _interopRequireDefault(_reactRelay);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let UserMentionTooltipRoute = (_temp = _class = class UserMentionTooltipRoute extends _reactRelay2.default.Route {}, _class.routeName = 'user-mention-tooltip-route', _class.queries = {
  repositoryOwner: Component => function (RQL_0) {
    return {
      calls: [{
        kind: 'Call',
        metadata: {
          type: 'String!'
        },
        name: 'login',
        value: {
          kind: 'CallVariable',
          callVariableName: 'username'
        }
      }],
      children: [].concat.apply([], [{
        fieldName: 'id',
        kind: 'Field',
        metadata: {
          isGenerated: true,
          isRequisite: true
        },
        type: 'ID'
      }, {
        fieldName: '__typename',
        kind: 'Field',
        metadata: {
          isGenerated: true,
          isRequisite: true
        },
        type: 'String'
      }, _reactRelay2.default.QL.__frag(RQL_0)]),
      fieldName: 'repositoryOwner',
      kind: 'Query',
      metadata: {
        isAbstract: true,
        identifyingArgName: 'login',
        identifyingArgType: 'String!'
      },
      name: 'User',
      type: 'RepositoryOwner'
    };
  }(Component.getFragment('repositoryOwner'))
}, _class.paramDefinitions = {
  username: { required: true }
}, _temp);
exports.default = UserMentionTooltipRoute;
