// settings-view/lib/error-view.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

/** @babel */
/** @jsx etch.dom */

var _etch = require('etch');

var _etch2 = _interopRequireDefault(_etch);

var ErrorView = (function () {
  function ErrorView(packageManager, _ref) {
    var _this = this;

    var message = _ref.message;
    var stderr = _ref.stderr;
    var packageInstallError = _ref.packageInstallError;

    _classCallCheck(this, ErrorView);

    _etch2['default'].initialize(this);

    this.isOutputHidden = true;
    this.refs.detailsArea.style.display = 'none';
    this.refs.details.textContent = stderr;
    this.refs.message.textContent = message;

    // Check for native build tools and show warning if missing.
    if (packageInstallError && process.platform === 'win32') {
      packageManager.checkNativeBuildTools()['catch'](function () {
        _this.refs.alert.appendChild(new CompileToolsErrorView().element);
      });
    }
  }

  _createClass(ErrorView, [{
    key: 'update',
    value: function update() {}
  }, {
    key: 'destroy',
    value: function destroy() {
      return _etch2['default'].destroy(this);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _etch2['default'].dom(
        'div',
        { className: 'error-message' },
        _etch2['default'].dom(
          'div',
          { ref: 'alert', className: 'alert alert-danger alert-dismissable native-key-bindings', tabIndex: '-1' },
          _etch2['default'].dom('button', { ref: 'close', className: 'close icon icon-x', onclick: function () {
              return _this2.destroy();
            } }),
          _etch2['default'].dom('span', { ref: 'message', className: 'native-key-bindings' }),
          _etch2['default'].dom(
            'a',
            { ref: 'detailsLink', className: 'alert-link error-link', onclick: function () {
                return _this2.toggleOutput();
              } },
            'Show output…'
          ),
          _etch2['default'].dom(
            'div',
            { ref: 'detailsArea', className: 'padded' },
            _etch2['default'].dom('pre', { ref: 'details', className: 'error-details text' })
          )
        )
      );
    }
  }, {
    key: 'toggleOutput',
    value: function toggleOutput() {
      if (this.isOutputHidden) {
        this.isOutputHidden = false;
        this.refs.detailsArea.style.display = '';
        this.refs.detailsLink.textContent = 'Hide output…';
      } else {
        this.isOutputHidden = true;
        this.refs.detailsArea.style.display = 'none';
        this.refs.detailsLink.textContent = 'Show output…';
      }
    }
  }]);

  return ErrorView;
})();

exports['default'] = ErrorView;

var CompileToolsErrorView = (function () {
  function CompileToolsErrorView() {
    _classCallCheck(this, CompileToolsErrorView);

    _etch2['default'].initialize(this);
  }

  _createClass(CompileToolsErrorView, [{
    key: 'update',
    value: function update() {}
  }, {
    key: 'render',
    value: function render() {
      return _etch2['default'].dom(
        'div',
        null,
        _etch2['default'].dom(
          'div',
          { className: 'icon icon-alert compile-tools-heading compile-tools-message' },
          'Compiler tools not found'
        ),
        _etch2['default'].dom(
          'div',
          { className: 'compile-tools-message' },
          'Packages that depend on modules that contain C/C++ code will fail to install.'
        ),
        _etch2['default'].dom(
          'div',
          { className: 'compile-tools-message' },
          _etch2['default'].dom(
            'span',
            null,
            'Read '
          ),
          _etch2['default'].dom(
            'a',
            { className: 'link', href: 'https://atom.io/docs/latest/build-instructions/windows' },
            'here'
          ),
          _etch2['default'].dom(
            'span',
            null,
            ' for instructions on installing Python and Visual Studio.'
          )
        ),
        _etch2['default'].dom(
          'div',
          { className: 'compile-tools-message' },
          _etch2['default'].dom(
            'span',
            null,
            'Run '
          ),
          _etch2['default'].dom(
            'code',
            { className: 'alert-danger' },
            'apm install --check'
          ),
          _etch2['default'].dom(
            'span',
            null,
            ' after installing to test compiling a native module.'
          )
        )
      );
    }
  }]);

  return CompileToolsErrorView;
})();

module.exports = exports['default'];
