// github/lib/models/repository-states/loading.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _state = require('./state');

var _state2 = _interopRequireDefault(_state);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

/**
 * Initial state to be used when it's uncertain whether or not a git repository is present in a working directory. If
 * it is a git repository, transition to Present, otherwise transition to Empty.
 */
let Loading = class Loading extends _state2.default {
  start() {
    var _this = this;

    return _asyncToGenerator(function* () {
      const dotGitDir = yield _this.resolveDotGitDir();
      if (dotGitDir) {
        _this.repository.setGitDirectoryPath(dotGitDir);
        const history = yield _this.loadHistoryPayload();
        return _this.transitionTo('Present', history);
      } else {
        return _this.transitionTo('Empty');
      }
    })();
  }

  isLoading() {
    return true;
  }

  init() {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      yield _this2.getLoadPromise();
      yield _this2.repository.init();
    })();
  }

  clone(remoteUrl) {
    var _this3 = this;

    return _asyncToGenerator(function* () {
      yield _this3.getLoadPromise();
      yield _this3.repository.clone(remoteUrl);
    })();
  }

  showGitTabLoading() {
    return true;
  }

  directResolveDotGitDir() {
    return this.git().resolveDotGitDir();
  }

  directGetConfig(key, options) {
    return this.git().getConfig(key, options);
  }

  directGetBlobContents(sha) {
    return this.git().getBlobContents(sha);
  }
};
exports.default = Loading;


_state2.default.register(Loading);
