// spell-check/lib/corrections-view.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i['return']) _i['return'](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError('Invalid attempt to destructure non-iterable instance'); } }; })();

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

/** @babel */

var _atomSelectList = require('atom-select-list');

var _atomSelectList2 = _interopRequireDefault(_atomSelectList);

var CorrectionsView = (function () {
  function CorrectionsView(editor, corrections, marker, updateTarget, updateCallback) {
    var _this = this;

    _classCallCheck(this, CorrectionsView);

    this.editor = editor;
    this.corrections = corrections;
    this.marker = marker;
    this.updateTarget = updateTarget;
    this.updateCallback = updateCallback;
    this.selectListView = new _atomSelectList2['default']({
      emptyMessage: 'No corrections',
      items: this.corrections,
      filterKeyForItem: function filterKeyForItem(item) {
        return item.label;
      },
      elementForItem: function elementForItem(item) {
        var element = document.createElement('li');
        if (item.isSuggestion) {
          // This is a word replacement suggestion.
          element.textContent = item.label;
        } else {
          // This is an operation such as add word.
          var em = document.createElement('em');
          em.textContent = item.label;
          element.appendChild(em);
        }
        return element;
      },
      didConfirmSelection: function didConfirmSelection(item) {
        _this.editor.transact(function () {
          if (item.isSuggestion) {
            // Update the buffer with the correction.
            _this.editor.setSelectedBufferRange(_this.marker.getBufferRange());
            _this.editor.insertText(item.suggestion);
          } else {
            // Build up the arguments object for this buffer and text.
            var projectPath = null;
            var relativePath = null;
            if (_this.editor && _this.editor.buffer && _this.editor.buffer.file && _this.editor.buffer.file.path) {
              var _atom$project$relativizePath = atom.project.relativizePath(_this.editor.buffer.file.path);

              var _atom$project$relativizePath2 = _slicedToArray(_atom$project$relativizePath, 2);

              projectPath = _atom$project$relativizePath2[0];
              relativePath = _atom$project$relativizePath2[1];
            }

            var args = { id: _this.id, projectPath: projectPath, relativePath: relativePath };
            // Send the "add" request to the plugin.
            item.plugin.add(args, item);
            // Update the buffer to handle the corrections.
            _this.updateCallback.bind(_this.updateTarget)();
          }
        });
        _this.destroy();
      },
      didConfirmEmptySelection: function didConfirmEmptySelection() {
        _this.destroy();
      },
      didCancelSelection: function didCancelSelection() {
        _this.destroy();
      }
    });
    this.selectListView.element.classList.add('spell-check-corrections', 'corrections', 'popover-list');
  }

  _createClass(CorrectionsView, [{
    key: 'attach',
    value: function attach() {
      var _this2 = this;

      this.previouslyFocusedElement = document.activeElement;
      this.overlayDecoration = this.editor.decorateMarker(this.marker, { type: 'overlay', item: this.selectListView });
      process.nextTick(function () {
        atom.views.readDocument(function () {
          _this2.selectListView.focus();
        });
      });
    }
  }, {
    key: 'destroy',
    value: _asyncToGenerator(function* () {
      if (!this.destroyed) {
        this.destroyed = true;
        this.overlayDecoration.destroy();
        yield this.selectListView.destroy();
        if (this.previouslyFocusedElement) {
          this.previouslyFocusedElement.focus();
          this.previouslyFocusedElement = null;
        }
      }
    })
  }]);

  return CorrectionsView;
})();

exports['default'] = CorrectionsView;
module.exports = exports['default'];
