// spell-check/lib/main.js
(function() {
  var SpellCheckView, Task, spellCheckViews,
    indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  Task = require('atom').Task;

  SpellCheckView = null;

  spellCheckViews = {};

  module.exports = {
    activate: function() {
      var handlerFilename, that;
      handlerFilename = require.resolve('./spell-check-handler');
      if (this.task == null) {
        this.task = new Task(handlerFilename);
      }
      that = this;
      this.task.on("spell-check:settings-changed", function(ignore) {
        return that.updateViews();
      });
      this.globalArgs = {
        locales: atom.config.get('spell-check.locales'),
        localePaths: atom.config.get('spell-check.localePaths'),
        useLocales: atom.config.get('spell-check.useLocales'),
        knownWords: atom.config.get('spell-check.knownWords'),
        addKnownWords: atom.config.get('spell-check.addKnownWords'),
        checkerPaths: []
      };
      this.sendGlobalArgs();
      atom.config.onDidChange('spell-check.locales', function(arg) {
        var newValue, oldValue;
        newValue = arg.newValue, oldValue = arg.oldValue;
        that.globalArgs.locales = newValue;
        return that.sendGlobalArgs();
      });
      atom.config.onDidChange('spell-check.localePaths', function(arg) {
        var newValue, oldValue;
        newValue = arg.newValue, oldValue = arg.oldValue;
        that.globalArgs.localePaths = newValue;
        return that.sendGlobalArgs();
      });
      atom.config.onDidChange('spell-check.useLocales', function(arg) {
        var newValue, oldValue;
        newValue = arg.newValue, oldValue = arg.oldValue;
        that.globalArgs.useLocales = newValue;
        return that.sendGlobalArgs();
      });
      atom.config.onDidChange('spell-check.knownWords', function(arg) {
        var newValue, oldValue;
        newValue = arg.newValue, oldValue = arg.oldValue;
        that.globalArgs.knownWords = newValue;
        return that.sendGlobalArgs();
      });
      atom.config.onDidChange('spell-check.addKnownWords', function(arg) {
        var newValue, oldValue;
        newValue = arg.newValue, oldValue = arg.oldValue;
        that.globalArgs.addKnownWords = newValue;
        return that.sendGlobalArgs();
      });
      this.commandSubscription = atom.commands.add('atom-workspace', {
        'spell-check:toggle': (function(_this) {
          return function() {
            return _this.toggle();
          };
        })(this)
      });
      this.viewsByEditor = new WeakMap;
      this.contextMenuEntries = [];
      return this.disposable = atom.workspace.observeTextEditors((function(_this) {
        return function(editor) {
          var editorId, spellCheckView;
          if (editor.largeFileMode) {
            return;
          }
          if (SpellCheckView == null) {
            SpellCheckView = require('./spell-check-view');
          }
          spellCheckView = new SpellCheckView(editor, _this.task, _this, function() {
            return _this.getInstance(_this.globalArgs);
          });
          editorId = editor.id;
          spellCheckViews[editorId] = {};
          spellCheckViews[editorId]['view'] = spellCheckView;
          spellCheckViews[editorId]['active'] = true;
          spellCheckViews[editorId]['editor'] = editor;
          return _this.viewsByEditor.set(editor, spellCheckView);
        };
      })(this));
    },
    deactivate: function() {
      var editorId, ref, ref1, view;
      if ((ref = this.instance) != null) {
        ref.deactivate();
      }
      this.instance = null;
      if ((ref1 = this.task) != null) {
        ref1.terminate();
      }
      this.task = null;
      this.commandSubscription.dispose();
      this.commandSubscription = null;
      for (editorId in spellCheckViews) {
        view = spellCheckViews[editorId].view;
        view.destroy();
      }
      spellCheckViews = {};
      this.viewsByEditor = new WeakMap;
      return this.disposable.dispose();
    },
    consumeSpellCheckers: function(checkerPaths) {
      var checkerPath, i, len, ref, ref1, results;
      if (!(checkerPaths instanceof Array)) {
        checkerPaths = [checkerPaths];
      }
      results = [];
      for (i = 0, len = checkerPaths.length; i < len; i++) {
        checkerPath = checkerPaths[i];
        if (indexOf.call(this.globalArgs.checkerPaths, checkerPath) < 0) {
          if ((ref = this.task) != null) {
            ref.send({
              type: "checker",
              checkerPath: checkerPath
            });
          }
          if ((ref1 = this.instance) != null) {
            ref1.addCheckerPath(checkerPath);
          }
          results.push(this.globalArgs.checkerPaths.push(checkerPath));
        } else {
          results.push(void 0);
        }
      }
      return results;
    },
    misspellingMarkersForEditor: function(editor) {
      return this.viewsByEditor.get(editor).markerLayer.getMarkers();
    },
    updateViews: function() {
      var editorId, results, view;
      results = [];
      for (editorId in spellCheckViews) {
        view = spellCheckViews[editorId];
        if (view['active']) {
          results.push(view['view'].updateMisspellings());
        } else {
          results.push(void 0);
        }
      }
      return results;
    },
    sendGlobalArgs: function() {
      return this.task.send({
        type: "global",
        global: this.globalArgs
      });
    },
    getInstance: function(globalArgs) {
      var SpellCheckerManager, checkerPath, i, len, ref;
      if (!this.instance) {
        SpellCheckerManager = require('./spell-check-manager');
        this.instance = SpellCheckerManager;
        this.instance.setGlobalArgs(globalArgs);
        ref = globalArgs.checkerPaths;
        for (i = 0, len = ref.length; i < len; i++) {
          checkerPath = ref[i];
          this.instance.addCheckerPath(checkerPath);
        }
      }
      return this.instance;
    },
    toggle: function() {
      var editorId;
      if (!atom.workspace.getActiveTextEditor()) {
        return;
      }
      editorId = atom.workspace.getActiveTextEditor().id;
      if (spellCheckViews[editorId]['active']) {
        spellCheckViews[editorId]['active'] = false;
        return spellCheckViews[editorId]['view'].unsubscribeFromBuffer();
      } else {
        spellCheckViews[editorId]['active'] = true;
        return spellCheckViews[editorId]['view'].subscribeToBuffer();
      }
    }
  };

}).call(this);
