// spell-check/lib/spell-check-task.js
(function() {
  var SpellCheckTask, idCounter;

  idCounter = 0;

  module.exports = SpellCheckTask = (function() {
    SpellCheckTask.handler = null;

    SpellCheckTask.callbacksById = {};

    function SpellCheckTask(task) {
      this.task = task;
      this.id = idCounter++;
    }

    SpellCheckTask.prototype.terminate = function() {
      return delete this.constructor.callbacksById[this.id];
    };

    SpellCheckTask.prototype.start = function(buffer) {
      var args, projectPath, ref, ref1, ref2, relativePath;
      projectPath = null;
      relativePath = null;
      if (buffer != null ? (ref = buffer.file) != null ? ref.path : void 0 : void 0) {
        ref1 = atom.project.relativizePath(buffer.file.path), projectPath = ref1[0], relativePath = ref1[1];
      }
      args = {
        id: this.id,
        projectPath: projectPath,
        relativePath: relativePath,
        text: buffer.getText()
      };
      return (ref2 = this.task) != null ? ref2.start(args, this.constructor.dispatchMisspellings) : void 0;
    };

    SpellCheckTask.prototype.onDidSpellCheck = function(callback) {
      return this.constructor.callbacksById[this.id] = callback;
    };

    SpellCheckTask.dispatchMisspellings = function(data) {
      var base, name;
      return typeof (base = SpellCheckTask.callbacksById)[name = data.id] === "function" ? base[name](data.misspellings) : void 0;
    };

    return SpellCheckTask;

  })();

}).call(this);
