// spell-check/lib/system-checker.js
(function() {
  var SystemChecker, spellchecker;

  spellchecker = require('spellchecker');

  SystemChecker = (function() {
    SystemChecker.prototype.spellchecker = null;

    SystemChecker.prototype.locale = null;

    SystemChecker.prototype.enabled = true;

    SystemChecker.prototype.reason = null;

    SystemChecker.prototype.paths = null;

    function SystemChecker(locale, paths) {
      this.locale = locale;
      this.paths = paths;
    }

    SystemChecker.prototype.deactivate = function() {};

    SystemChecker.prototype.getId = function() {
      return "spell-check:" + this.locale.toLowerCase().replace("_", "-");
    };

    SystemChecker.prototype.getName = function() {
      return "System Dictionary (" + this.locale + ")";
    };

    SystemChecker.prototype.getPriority = function() {
      return 100;
    };

    SystemChecker.prototype.isEnabled = function() {
      return this.enabled;
    };

    SystemChecker.prototype.getStatus = function() {
      if (this.enabled) {
        return "Working correctly.";
      } else {
        return this.reason;
      }
    };

    SystemChecker.prototype.providesSpelling = function(args) {
      return true;
    };

    SystemChecker.prototype.providesSuggestions = function(args) {
      return true;
    };

    SystemChecker.prototype.providesAdding = function(args) {
      return false;
    };

    SystemChecker.prototype.check = function(args, text) {
      this.deferredInit();
      return {
        invertIncorrectAsCorrect: true,
        incorrect: this.spellchecker.checkSpelling(text)
      };
    };

    SystemChecker.prototype.suggest = function(args, word) {
      this.deferredInit();
      return this.spellchecker.getCorrectionsForMisspelling(word);
    };

    SystemChecker.prototype.deferredInit = function() {
      var i, len, path, ref, vendor;
      if (this.spellchecker) {
        return;
      }
      this.spellchecker = new spellchecker.Spellchecker;
      if (/win32/.test(process.platform)) {
        if (this.spellchecker.setDictionary(this.locale, "C:\\")) {
          return;
        }
      }
      ref = this.paths;
      for (i = 0, len = ref.length; i < len; i++) {
        path = ref[i];
        if (this.spellchecker.setDictionary(this.locale, path)) {
          return;
        }
      }
      if (/linux/.test(process.platform)) {
        if (this.spellchecker.setDictionary(this.locale, "/usr/share/hunspell")) {
          return;
        }
        if (this.spellchecker.setDictionary(this.locale, "/usr/share/myspell")) {
          return;
        }
        if (this.spellchecker.setDictionary(this.locale, "/usr/share/myspell/dicts")) {
          return;
        }
      }
      if (/darwin/.test(process.platform)) {
        if (this.spellchecker.setDictionary(this.locale, "/")) {
          return;
        }
        if (this.spellchecker.setDictionary(this.locale, "/System/Library/Spelling")) {
          return;
        }
      }
      path = require('path');
      vendor = path.join(__dirname, "..", "node_modules", "spellchecker", "vendor", "hunspell_dictionaries");
      if (this.spellchecker.setDictionary(this.locale, vendor)) {
        return;
      }
      this.enabled = false;
      this.reason = "Cannot find dictionary for " + this.locale + ".";
      return console.log(this.getId(), "Can't load " + this.locale + ": " + this.reason);
    };

    return SystemChecker;

  })();

  module.exports = SystemChecker;

}).call(this);
