// github/lib/containers/issueish-lookup-by-number-container.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IssueishLookupByNumber = undefined;

var _desc, _value, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRelay = require('react-relay');

var _reactRelay2 = _interopRequireDefault(_reactRelay);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _coreDecorators = require('core-decorators');

var _issueishPaneItemContainer = require('./issueish-pane-item-container');

var _issueishPaneItemContainer2 = _interopRequireDefault(_issueishPaneItemContainer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

let IssueishLookupByNumber = exports.IssueishLookupByNumber = (_class = (_temp = _class2 = class IssueishLookupByNumber extends _react2.default.Component {

  componentDidMount() {
    this.updateTitle();
  }

  componentDidUpdate() {
    this.updateTitle();
  }

  updateTitle() {
    const repository = this.props.query.repository;

    if (repository && repository.issueish) {
      var _props$relay$variable = this.props.relay.variables;
      const repoOwner = _props$relay$variable.repoOwner,
            repoName = _props$relay$variable.repoName,
            issueishNumber = _props$relay$variable.issueishNumber;
      // eslint-disable-next-line no-irregular-whitespace

      let title = `${repoOwner}/${repoName}#${issueishNumber} — ${repository.issueish.title}`;
      if (repository.issueish.__typename === 'Issue') {
        title = 'Issue: ' + title;
      } else {
        title = 'PR: ' + title;
      }
      this.props.onTitleChange(title);
    }
  }

  render() {
    const repo = this.props.query.repository;
    if (!repo || !repo.issueish) {
      return _react2.default.createElement(
        'div',
        null,
        'Issue/PR #',
        this.props.relay.variables.issueishNumber,
        ' not found'
      ); // TODO: no PRs
    }
    return _react2.default.createElement(_issueishPaneItemContainer2.default, { repository: repo, issueish: repo.issueish, switchToIssueish: this.switchToIssueish });
  }

  switchToIssueish(repoOwner, repoName, issueishNumber) {
    this.props.relay.setVariables({ repoOwner, repoName, issueishNumber });
  }
}, _class2.propTypes = {
  onTitleChange: _propTypes2.default.func.isRequired,
  query: _propTypes2.default.shape({
    repository: _propTypes2.default.object
  }),
  relay: _propTypes2.default.shape({
    setVariables: _propTypes2.default.func.isRequired,
    variables: _propTypes2.default.shape({
      repoOwner: _propTypes2.default.string.isRequired,
      repoName: _propTypes2.default.string.isRequired,
      issueishNumber: _propTypes2.default.number.isRequired
    }).isRequired
  }).isRequired
}, _temp), (_applyDecoratedDescriptor(_class.prototype, 'switchToIssueish', [_coreDecorators.autobind], Object.getOwnPropertyDescriptor(_class.prototype, 'switchToIssueish'), _class.prototype)), _class);
exports.default = _reactRelay2.default.createContainer(IssueishLookupByNumber, {
  initialVariables: {
    repoOwner: null,
    repoName: null,
    issueishNumber: null
  },

  // TOOD: remove issueish/PR hack once GraphQL is fixed
  fragments: {
    query: () => function (RQL_0, RQL_1, RQL_2) {
      return {
        children: [{
          calls: [{
            kind: 'Call',
            metadata: {},
            name: 'owner',
            value: {
              kind: 'CallVariable',
              callVariableName: 'repoOwner'
            }
          }, {
            kind: 'Call',
            metadata: {},
            name: 'name',
            value: {
              kind: 'CallVariable',
              callVariableName: 'repoName'
            }
          }],
          children: [].concat.apply([], [{
            alias: 'issueish',
            calls: [{
              kind: 'Call',
              metadata: {},
              name: 'number',
              value: {
                kind: 'CallVariable',
                callVariableName: 'issueishNumber'
              }
            }],
            children: [{
              fieldName: '__typename',
              kind: 'Field',
              metadata: {
                isRequisite: true
              },
              type: 'String'
            }, {
              children: [].concat.apply([], [{
                fieldName: 'title',
                kind: 'Field',
                metadata: {},
                type: 'String'
              }, {
                fieldName: 'id',
                kind: 'Field',
                metadata: {
                  isGenerated: true,
                  isRequisite: true
                },
                type: 'ID'
              }, _reactRelay2.default.QL.__frag(RQL_1)]),
              id: _reactRelay2.default.QL.__id(),
              kind: 'Fragment',
              metadata: {},
              name: 'Issue',
              type: 'Issue'
            }, {
              children: [].concat.apply([], [{
                fieldName: 'title',
                kind: 'Field',
                metadata: {},
                type: 'String'
              }, {
                fieldName: 'id',
                kind: 'Field',
                metadata: {
                  isGenerated: true,
                  isRequisite: true
                },
                type: 'ID'
              }, _reactRelay2.default.QL.__frag(RQL_2)]),
              id: _reactRelay2.default.QL.__id(),
              kind: 'Fragment',
              metadata: {},
              name: 'PullRequest',
              type: 'PullRequest'
            }, {
              children: [{
                fieldName: 'id',
                kind: 'Field',
                metadata: {
                  isGenerated: true,
                  isRequisite: true
                },
                type: 'ID'
              }, {
                fieldName: '__typename',
                kind: 'Field',
                metadata: {
                  isGenerated: true,
                  isRequisite: true
                },
                type: 'String'
              }],
              id: _reactRelay2.default.QL.__id(),
              kind: 'Fragment',
              metadata: {
                isAbstract: true
              },
              name: 'IdFragment',
              type: 'Node'
            }],
            fieldName: 'issueOrPullRequest',
            kind: 'Field',
            metadata: {
              canHaveSubselections: true,
              inferredRootCallName: 'node',
              inferredPrimaryKey: 'id',
              isAbstract: true
            },
            type: 'IssueOrPullRequest'
          }, {
            fieldName: 'id',
            kind: 'Field',
            metadata: {
              isGenerated: true,
              isRequisite: true
            },
            type: 'ID'
          }, _reactRelay2.default.QL.__frag(RQL_0)]),
          fieldName: 'repository',
          kind: 'Field',
          metadata: {
            canHaveSubselections: true,
            inferredRootCallName: 'node',
            inferredPrimaryKey: 'id'
          },
          type: 'Repository'
        }],
        id: _reactRelay2.default.QL.__id(),
        kind: 'Fragment',
        metadata: {},
        name: 'Issueish_QueryRelayQL',
        type: 'Query'
      };
    }(_issueishPaneItemContainer2.default.getFragment('repository'), _issueishPaneItemContainer2.default.getFragment('issueish'), _issueishPaneItemContainer2.default.getFragment('issueish'))
  }
});
