// github/lib/containers/pr-selection-by-url-container.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PrSelectionByUrl = undefined;

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRelay = require('react-relay');

var _reactRelay2 = _interopRequireDefault(_reactRelay);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _prInfoContainer = require('./pr-info-container');

var _prInfoContainer2 = _interopRequireDefault(_prInfoContainer);

var _prUrlInputBox = require('../views/pr-url-input-box');

var _prUrlInputBox2 = _interopRequireDefault(_prUrlInputBox);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let PrSelectionByUrl = exports.PrSelectionByUrl = (_temp = _class = class PrSelectionByUrl extends _react2.default.Component {

  render() {
    // TODO: render a selector if multiple PRs
    const resource = this.props.query && this.props.query.resource;
    if (!resource || resource.__typename !== 'PullRequest') {
      // return <div>PR not found</div>;
      return _react2.default.createElement(
        'div',
        { className: 'github-PrUrlInputBox-Container' },
        _react2.default.createElement(
          _prUrlInputBox2.default,
          { onSubmit: this.props.onSelectPr },
          _react2.default.createElement(
            'p',
            null,
            _react2.default.createElement(
              'span',
              { style: { display: 'block' } },
              'This branch is pinned to the pull request at this URL:'
            ),
            _react2.default.createElement('input', { value: this.props.prUrl, style: { width: '100%' }, readOnly: true }),
            _react2.default.createElement(
              'span',
              { style: { display: 'block' } },
              'but we couldn\'t find a pull request at that URL.'
            )
          ),
          _react2.default.createElement(
            'p',
            null,
            'You can manually pin another GitHub pull request to the current branch by entering its URL:'
          )
        )
      );
    }
    return _react2.default.createElement(_prInfoContainer2.default, {
      pullRequest: resource,
      pinnedByUrl: true,
      onUnpinPr: this.props.onUnpinPr
    });
  }

  setPr(prLink) {
    this.props.onSelectPr(prLink);
  }
}, _class.propTypes = {
  query: _propTypes2.default.shape({
    resource: _propTypes2.default.object
  }),
  prUrl: _propTypes2.default.string,
  onSelectPr: _propTypes2.default.func.isRequired,
  onUnpinPr: _propTypes2.default.func.isRequired
}, _temp);
exports.default = _reactRelay2.default.createContainer(PrSelectionByUrl, {
  initialVariables: {
    prUrl: null
  },

  fragments: {
    query: () => function (RQL_0) {
      return {
        children: [{
          calls: [{
            kind: 'Call',
            metadata: {
              type: 'URI!'
            },
            name: 'url',
            value: {
              kind: 'CallVariable',
              callVariableName: 'prUrl'
            }
          }],
          children: [{
            fieldName: '__typename',
            kind: 'Field',
            metadata: {
              isRequisite: true
            },
            type: 'String'
          }, {
            children: [].concat.apply([], [{
              fieldName: 'id',
              kind: 'Field',
              metadata: {
                isGenerated: true,
                isRequisite: true
              },
              type: 'ID'
            }, _reactRelay2.default.QL.__frag(RQL_0)]),
            id: _reactRelay2.default.QL.__id(),
            kind: 'Fragment',
            metadata: {},
            name: 'PullRequest',
            type: 'PullRequest'
          }, {
            children: [{
              fieldName: 'id',
              kind: 'Field',
              metadata: {
                isGenerated: true,
                isRequisite: true
              },
              type: 'ID'
            }, {
              fieldName: '__typename',
              kind: 'Field',
              metadata: {
                isGenerated: true,
                isRequisite: true
              },
              type: 'String'
            }],
            id: _reactRelay2.default.QL.__id(),
            kind: 'Fragment',
            metadata: {
              isAbstract: true
            },
            name: 'IdFragment',
            type: 'Node'
          }],
          fieldName: 'resource',
          kind: 'Field',
          metadata: {
            canHaveSubselections: true,
            inferredRootCallName: 'node',
            inferredPrimaryKey: 'id',
            isAbstract: true
          },
          type: 'UniformResourceLocatable'
        }],
        id: _reactRelay2.default.QL.__id(),
        kind: 'Fragment',
        metadata: {},
        name: 'Pr_QueryRelayQL',
        type: 'Query'
      };
    }(_prInfoContainer2.default.getFragment('pullRequest'))
  }
});
