// github/lib/containers/relay-root-container.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = RelayRootContainer;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRelay = require('react-relay');

var _reactRelay2 = _interopRequireDefault(_reactRelay);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This is a copy of `RelayRootContainer` from the Relay
 * codebase, but with an optional `environment` prop
 * that can be used instead of the singleton `RelayStore`.
 */
function RelayRootContainer(_ref) {
  let Component = _ref.Component,
      forceFetch = _ref.forceFetch,
      onReadyStateChange = _ref.onReadyStateChange,
      renderFailure = _ref.renderFailure,
      renderFetched = _ref.renderFetched,
      renderLoading = _ref.renderLoading,
      route = _ref.route,
      shouldFetch = _ref.shouldFetch,
      environment = _ref.environment;

  return _react2.default.createElement(_reactRelay2.default.Renderer, {
    Container: Component,
    forceFetch: forceFetch,
    onReadyStateChange: onReadyStateChange,
    queryConfig: route,
    environment: environment || _reactRelay2.default.Store,
    shouldFetch: shouldFetch,
    render: (_ref2) => {
      let done = _ref2.done,
          error = _ref2.error,
          props = _ref2.props,
          retry = _ref2.retry,
          stale = _ref2.stale;

      if (error) {
        if (renderFailure) {
          return renderFailure(error, retry);
        }
      } else if (props) {
        if (renderFetched) {
          return renderFetched(props, { done, stale });
        } else {
          return _react2.default.createElement(Component, props);
        }
      } else {
        if (renderLoading) {
          return renderLoading();
        }
      }
      return undefined;
    }
  });
}

RelayRootContainer.propTypes = {
  Component: _reactRelay2.default.PropTypes.Container,
  forceFetch: _propTypes2.default.bool,
  onReadyStateChange: _propTypes2.default.func,
  renderFailure: _propTypes2.default.func,
  renderFetched: _propTypes2.default.func,
  renderLoading: _propTypes2.default.func,
  route: _reactRelay2.default.PropTypes.QueryConfig.isRequired,
  shouldFetch: _propTypes2.default.bool,
  environment: _propTypes2.default.object
};
