// github/lib/containers/user-mention-tooltip-container.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _class, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRelay = require('react-relay');

var _reactRelay2 = _interopRequireDefault(_reactRelay);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _octicon = require('../views/octicon');

var _octicon2 = _interopRequireDefault(_octicon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let UserMentionTooltip = (_temp = _class = class UserMentionTooltip extends _react2.default.Component {

  render() {
    const owner = this.props.repositoryOwner;
    const login = owner.login,
          company = owner.company,
          repositories = owner.repositories,
          members = owner.members;

    return _react2.default.createElement(
      'div',
      { className: 'github-UserMentionTooltip' },
      _react2.default.createElement(
        'div',
        { className: 'github-UserMentionTooltip-avatar' },
        _react2.default.createElement('img', { src: owner.avatarUrl })
      ),
      _react2.default.createElement(
        'div',
        { className: 'github-UserMentionTooltip-info' },
        _react2.default.createElement(
          'div',
          { className: 'github-UserMentionTooltip-info-username' },
          _react2.default.createElement(_octicon2.default, { icon: 'mention' }),
          _react2.default.createElement(
            'strong',
            null,
            login
          )
        ),
        company && _react2.default.createElement(
          'div',
          null,
          _react2.default.createElement(_octicon2.default, { icon: 'briefcase' }),
          _react2.default.createElement(
            'span',
            null,
            company
          )
        ),
        members && _react2.default.createElement(
          'div',
          null,
          _react2.default.createElement(_octicon2.default, { icon: 'organization' }),
          _react2.default.createElement(
            'span',
            null,
            members.totalCount,
            ' members'
          )
        ),
        _react2.default.createElement(
          'div',
          null,
          _react2.default.createElement(_octicon2.default, { icon: 'repo' }),
          _react2.default.createElement(
            'span',
            null,
            repositories.totalCount,
            ' repositories'
          )
        )
      ),
      _react2.default.createElement('div', { style: { clear: 'both' } })
    );
  }
}, _class.propTypes = {
  repositoryOwner: _propTypes2.default.shape({
    login: _propTypes2.default.string.isRequired,
    avatarUrl: _propTypes2.default.string.isRequired,
    repositories: _propTypes2.default.shape({
      totalCount: _propTypes2.default.number.isRequired
    }).isRequired,

    // Users
    company: _propTypes2.default.string,

    // Organizations
    members: _propTypes2.default.shape({
      totalCount: _propTypes2.default.number.isRequired
    })
  }).isRequired
}, _temp);
exports.default = _reactRelay2.default.createContainer(UserMentionTooltip, {
  fragments: {
    repositoryOwner: () => function () {
      return {
        children: [{
          fieldName: 'login',
          kind: 'Field',
          metadata: {},
          type: 'String'
        }, {
          fieldName: 'avatarUrl',
          kind: 'Field',
          metadata: {},
          type: 'URI'
        }, {
          children: [{
            fieldName: 'totalCount',
            kind: 'Field',
            metadata: {},
            type: 'Int'
          }],
          fieldName: 'repositories',
          kind: 'Field',
          metadata: {
            canHaveSubselections: true,
            isConnection: true
          },
          type: 'RepositoryConnection'
        }, {
          fieldName: 'id',
          kind: 'Field',
          metadata: {
            isGenerated: true,
            isRequisite: true
          },
          type: 'ID'
        }, {
          fieldName: '__typename',
          kind: 'Field',
          metadata: {
            isGenerated: true,
            isRequisite: true
          },
          type: 'String'
        }, {
          children: [{
            fieldName: 'company',
            kind: 'Field',
            metadata: {},
            type: 'String'
          }, {
            fieldName: 'id',
            kind: 'Field',
            metadata: {
              isGenerated: true,
              isRequisite: true
            },
            type: 'ID'
          }],
          id: _reactRelay2.default.QL.__id(),
          kind: 'Fragment',
          metadata: {},
          name: 'User',
          type: 'User'
        }, {
          children: [{
            children: [{
              fieldName: 'totalCount',
              kind: 'Field',
              metadata: {},
              type: 'Int'
            }],
            fieldName: 'members',
            kind: 'Field',
            metadata: {
              canHaveSubselections: true,
              isConnection: true
            },
            type: 'UserConnection'
          }, {
            fieldName: 'id',
            kind: 'Field',
            metadata: {
              isGenerated: true,
              isRequisite: true
            },
            type: 'ID'
          }],
          id: _reactRelay2.default.QL.__id(),
          kind: 'Fragment',
          metadata: {},
          name: 'Organization',
          type: 'Organization'
        }],
        id: _reactRelay2.default.QL.__id(),
        kind: 'Fragment',
        metadata: {
          isAbstract: true
        },
        name: 'User_RepositoryOwnerRelayQL',
        type: 'RepositoryOwner'
      };
    }()
  }
});
