// github/lib/atom-items/issueish-tooltip-item.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _issueishTooltipRoute = require('../routes/issueish-tooltip-route');

var _issueishTooltipRoute2 = _interopRequireDefault(_issueishTooltipRoute);

var _issueishTooltipContainer = require('../containers/issueish-tooltip-container');

var _issueishTooltipContainer2 = _interopRequireDefault(_issueishTooltipContainer);

var _relayRootContainer = require('../containers/relay-root-container');

var _relayRootContainer2 = _interopRequireDefault(_relayRootContainer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let IssueishTooltipItem = class IssueishTooltipItem {
  constructor(issueishUrl, relayEnvironment) {
    this.issueishUrl = issueishUrl;
    this.relayEnvironment = relayEnvironment;
  }

  getElement() {
    return this.element;
  }

  get element() {
    if (!this._element) {
      const route = new _issueishTooltipRoute2.default({ issueishUrl: this.issueishUrl });
      this._element = document.createElement('div');
      const rootContainer = _react2.default.createElement(_relayRootContainer2.default, {
        Component: _issueishTooltipContainer2.default,
        route: route,
        environment: this.relayEnvironment,
        renderLoading: () => {
          // TODO: adjust min width/height
          return _react2.default.createElement(
            'div',
            { className: 'github-Loader' },
            _react2.default.createElement('span', { className: 'github-Spinner' })
          );
        },
        renderFailure: () => _react2.default.createElement(
          'div',
          null,
          'Could not load information'
        )
      });
      this._component = _reactDom2.default.render(rootContainer, this._element);
    }

    return this._element;
  }

  destroy() {
    if (this._element) {
      _reactDom2.default.unmountComponentAtNode(this._element);
      delete this._element;
    }
  }
};
exports.default = IssueishTooltipItem;
