// symbols-view/lib/project-view.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/** @babel */

var _atom = require('atom');

var _humanizePlus = require('humanize-plus');

var _humanizePlus2 = _interopRequireDefault(_humanizePlus);

var _symbolsView = require('./symbols-view');

var _symbolsView2 = _interopRequireDefault(_symbolsView);

var _tagReader = require('./tag-reader');

var _tagReader2 = _interopRequireDefault(_tagReader);

var _getTagsFile = require('./get-tags-file');

var _getTagsFile2 = _interopRequireDefault(_getTagsFile);

var ProjectView = (function (_SymbolsView) {
  _inherits(ProjectView, _SymbolsView);

  function ProjectView(stack) {
    _classCallCheck(this, ProjectView);

    _get(Object.getPrototypeOf(ProjectView.prototype), 'constructor', this).call(this, stack, 'Project has no tags file or it is empty', 10);
    this.reloadTags = true;
  }

  _createClass(ProjectView, [{
    key: 'destroy',
    value: function destroy() {
      this.stopTask();
      this.unwatchTagsFiles();
      return _get(Object.getPrototypeOf(ProjectView.prototype), 'destroy', this).call(this);
    }
  }, {
    key: 'toggle',
    value: function toggle() {
      if (this.panel.isVisible()) {
        this.cancel();
      } else {
        this.populate();
        this.attach();
      }
    }
  }, {
    key: 'populate',
    value: _asyncToGenerator(function* () {
      var _this = this;

      if (this.tags) {
        yield this.selectListView.update({ items: this.tags });
      }

      if (this.reloadTags) {
        this.reloadTags = false;
        this.startTask();

        if (this.tags) {
          yield this.selectListView.update({
            loadingMessage: 'Reloading project symbols…'
          });
        } else {
          yield* (function* () {
            yield _this.selectListView.update({
              loadingMessage: 'Loading project symbols…',
              loadingBadge: 0
            });
            var tagsRead = 0;
            _this.loadTagsTask.on('tags', function (tags) {
              tagsRead += tags.length;
              _this.selectListView.update({ loadingBadge: _humanizePlus2['default'].intComma(tagsRead) });
            });
          })();
        }
      }
    })
  }, {
    key: 'stopTask',
    value: function stopTask() {
      if (this.loadTagsTask) {
        this.loadTagsTask.terminate();
      }
    }
  }, {
    key: 'startTask',
    value: function startTask() {
      var _this2 = this;

      this.stopTask();

      this.loadTagsTask = _tagReader2['default'].getAllTags(function (tags) {
        _this2.tags = tags;
        _this2.reloadTags = _this2.tags.length === 0;
        _this2.selectListView.update({
          loadingMessage: null,
          loadingBadge: null,
          items: _this2.tags
        });
      });

      this.watchTagsFiles();
    }
  }, {
    key: 'watchTagsFiles',
    value: function watchTagsFiles() {
      var _this3 = this;

      this.unwatchTagsFiles();

      this.tagsFileSubscriptions = new _atom.CompositeDisposable();
      var reloadTags = function reloadTags() {
        _this3.reloadTags = true;
        _this3.watchTagsFiles();
      };

      for (var projectPath of Array.from(atom.project.getPaths())) {
        var tagsFilePath = (0, _getTagsFile2['default'])(projectPath);
        if (tagsFilePath) {
          var tagsFile = new _atom.File(tagsFilePath);
          this.tagsFileSubscriptions.add(tagsFile.onDidChange(reloadTags));
          this.tagsFileSubscriptions.add(tagsFile.onDidDelete(reloadTags));
          this.tagsFileSubscriptions.add(tagsFile.onDidRename(reloadTags));
        }
      }
    }
  }, {
    key: 'unwatchTagsFiles',
    value: function unwatchTagsFiles() {
      if (this.tagsFileSubscriptions) {
        this.tagsFileSubscriptions.dispose();
      }
      this.tagsFileSubscriptions = null;
    }
  }]);

  return ProjectView;
})(_symbolsView2['default']);

exports['default'] = ProjectView;
module.exports = exports['default'];
