// timecop/lib/cache-panel-view.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

/** @babel */
/** @jsx etch.dom */

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _etch = require('etch');

var _etch2 = _interopRequireDefault(_etch);

var CachePanelView = (function () {
  function CachePanelView() {
    _classCallCheck(this, CachePanelView);

    _etch2['default'].initialize(this);
  }

  _createClass(CachePanelView, [{
    key: 'update',
    value: function update() {}
  }, {
    key: 'destroy',
    value: function destroy() {
      return _etch2['default'].destroy(this);
    }
  }, {
    key: 'render',
    value: function render() {
      return _etch2['default'].dom(
        'div',
        { className: 'tool-panel padded package-panel' },
        _etch2['default'].dom(
          'div',
          { className: 'inset-panel' },
          _etch2['default'].dom(
            'div',
            { className: 'panel-heading' },
            'Compile Cache'
          ),
          _etch2['default'].dom(
            'div',
            { className: 'panel-body padded' },
            _etch2['default'].dom(
              'div',
              { className: 'timing' },
              _etch2['default'].dom(
                'span',
                { className: 'inline-block' },
                'CoffeeScript files compiled'
              ),
              _etch2['default'].dom(
                'span',
                { className: 'inline-block highlight-info', ref: 'coffeeCompileCount' },
                '0'
              )
            ),
            _etch2['default'].dom(
              'div',
              { className: 'timing' },
              _etch2['default'].dom(
                'span',
                { className: 'inline-block' },
                'Babel files compiled'
              ),
              _etch2['default'].dom('span', { className: 'inline-block highlight-info', ref: 'babelCompileCount' })
            ),
            _etch2['default'].dom(
              'div',
              { className: 'timing' },
              _etch2['default'].dom(
                'span',
                { className: 'inline-block' },
                'Typescript files compiled'
              ),
              _etch2['default'].dom('span', { className: 'inline-block highlight-info', ref: 'typescriptCompileCount' })
            ),
            _etch2['default'].dom(
              'div',
              { className: 'timing' },
              _etch2['default'].dom(
                'span',
                { className: 'inline-block' },
                'CSON files compiled'
              ),
              _etch2['default'].dom('span', { className: 'inline-block highlight-info', ref: 'csonCompileCount' })
            ),
            _etch2['default'].dom(
              'div',
              { className: 'timing' },
              _etch2['default'].dom(
                'span',
                { className: 'inline-block' },
                'Less files compiled'
              ),
              _etch2['default'].dom('span', { className: 'inline-block highlight-info', ref: 'lessCompileCount' })
            )
          )
        )
      );
    }
  }, {
    key: 'populate',
    value: function populate() {
      var compileCacheStats = this.getCompileCacheStats();
      if (compileCacheStats) {
        this.refs.coffeeCompileCount.textContent = compileCacheStats['.coffee'].misses;
        this.refs.babelCompileCount.textContent = compileCacheStats['.js'].misses;
        this.refs.typescriptCompileCount.textContent = compileCacheStats['.ts'].misses;
      }

      this.refs.csonCompileCount.textContent = this.getCsonCompiles();
      this.refs.lessCompileCount.textContent = this.getLessCompiles();
    }
  }, {
    key: 'getCompileCacheStats',
    value: function getCompileCacheStats() {
      try {
        return require(_path2['default'].join(atom.getLoadSettings().resourcePath, 'src', 'compile-cache')).getCacheStats();
      } catch (error) {
        return null;
      }
    }
  }, {
    key: 'getCsonCompiles',
    value: function getCsonCompiles() {
      try {
        var CSON = require(_path2['default'].join(atom.getLoadSettings().resourcePath, 'node_modules', 'season'));
        if (CSON.getCacheMisses) {
          return CSON.getCacheMisses() || 0;
        } else {
          return 0;
        }
      } catch (error) {
        return 0;
      }
    }
  }, {
    key: 'getLessCompiles',
    value: function getLessCompiles() {
      var lessCache = atom.themes.lessCache;
      if (lessCache && lessCache.cache && lessCache.cache.stats && lessCache.cache.stats.misses) {
        return lessCache.cache.stats.misses || 0;
      } else {
        return 0;
      }
    }
  }]);

  return CachePanelView;
})();

exports['default'] = CachePanelView;
module.exports = exports['default'];
